{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "072b2b01-ee39-495c-8e0a-6c3ad7f7dc4f",
   "metadata": {
    "tags": []
   },
   "source": [
    "# M1 MEG - UE 5 A2B - TP5\n",
    "## Pipeline d'analyse RNA-seq - Partie 1\n",
    "\n",
    "L'objectif de ce TP est de faire tourner les premières étapes du pipeline d'analyse RNA-seq : \n",
    "- Comprendre les différentes étapes\n",
    "- Comprendre le lien entre les étapes\n",
    "- Analyser les résultats\n",
    "\n",
    "**Quelques rappels** : \n",
    "- Un ordinateur / serveur de calcul est complètement stupide : il fera *exactement* ce que vous avez demandé de faire (ni plus ni moins).\n",
    "- 99% des problèmes en informatique sont situés entre l'écran et la chaise.\n",
    "\n",
    "**Mode de fonctionnement:**\n",
    "\n",
    "- Les commandes que vous aurez à executer vous sont données dans le notebook. \n",
    "- À vous de les exécuter dans l'ordre\n",
    "- Inutile de tout faire tourner d'un seul coup, un objectif essentiel est de *comprendre* ce que font les programmes, ce qu'ils retournent et *analyser* les résultats.\n",
    "\n",
    "**Emplacements des Données :**\n",
    "- Données Brutes : `/srv/data/meg-m1-a2b/blumenthal-2014/chr7/1-fastq/`\n",
    "- Annotation du Génome : `/srv/data/Genomes/Mmu/GRCm39/extracted/genome_annotation-M37.gtf`\n",
    "- Index du Génome : `/srv/data/Genomes/Mmu/GRCm39/indexes_upto49bases/`\n",
    "\n",
    "**Configuration de l'environnement**\n",
    "Les outils requis sont pré-installés dans l'environnement meg-m1-ue5-unix2 dans lequel vous êtes :\n",
    "- FastQC (v0.12.1) - Contrôle qualité\n",
    "- MultiQC (v1.13) - Rapports agrégés\n",
    "- fastp (v0.23.1) - Prétraitement des reads\n",
    "- STAR (v2.7.11a) - Alignement des reads\n",
    "- samtools (v1.18) - Manipulation des fichiers BAM\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "70a8e542",
   "metadata": {},
   "source": [
    "# 1. Prépration\n",
    "\n",
    "## Création des répertoires de sortie\n",
    "\n",
    "Comme indiqué ci-dessus, un ordinateur est complètement stupide et ne fera que ce que vous lui demandez de faire. \n",
    "Par exemple, si vous demandez à un programme d'enregistrer des fichiers de sortie dans un dossier *qui n'existe pas*, il va y avoir un soucis.\n",
    "\n",
    "Il est donc important de *préparer* les dossiers de sortie et de les créer.\n",
    "\n",
    "Nous allons créer un dossier principal où tous les résultats seront enregistrés, puis des sous-dossiers pour chaque étape / outil.\n",
    "\n",
    "**Conseil** : les meilleurs amis d'un.e bioinformaticien.ne reste le papier et le stylo : ayez toujours une feuille / un cahier à côté de vous où vous allez *plannifier* vos analyses : quelles étapes, quels outils, etc..."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "c9b06970",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "mkdir -p Analysis_chr7/1-beforeTrimming/fastqc\n",
    "mkdir -p Analysis_chr7/1-beforeTrimming/multiqc\n",
    "\n",
    "mkdir -p Analysis_chr7/2-afterTrimming/fastp\n",
    "mkdir -p Analysis_chr7/2-afterTrimming/fastqc\n",
    "mkdir -p Analysis_chr7/2-afterTrimming/multiqc\n",
    "\n",
    "mkdir -p Analysis_chr7/3-mapping/bam\n",
    "mkdir -p Analysis_chr7/3-mapping/multiqc"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2aab7f37",
   "metadata": {},
   "source": [
    "On a créé un dossier principal (`Analysis_chr7`), trois sous-dossiers (`1-beforeTrimming`, `2-afterTrimming`et `3-mapping`), ainsi que des dossier pour chaque outil.\n",
    "\n",
    "**Questions** : \n",
    "- À quoi sert l'option `-p` du programme `mkdir` ?\n",
    "- Quel peut être l'intérêt de numéroter les dossiers ?"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8b284ef6-f793-4df1-8013-27c432f2e611",
   "metadata": {},
   "source": [
    "## Vérification des fichiers de données initiaux\n",
    "\n",
    "Grâce aux sommes de contrôle, on peut vérifier l'intégrité de nos données de base."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "180809e8-cd56-4d23-82f7-d50ee692a768",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "total 481M\n",
      "-rw-rw-r--+ 1     1003 1012 372 Sep  9 15:47 md5sum.txt\n",
      "-rw-rw-r--+ 1     1003 1012 14K Sep  3 17:18 samples-Blumenthal2014.tsv\n",
      "-rw-rw-r--+ 1 yclement 1289  33 Sep 18 13:43 samples.chr7.txt\n",
      "-rw-rw-r--+ 1     1003 1012 80M Sep  9 15:46 SRX1589831_chr7_R1.fastq.gz\n",
      "-rw-rw-r--+ 1     1003 1012 82M Sep  9 15:46 SRX1589831_chr7_R2.fastq.gz\n",
      "-rw-rw-r--+ 1     1003 1012 90M Sep  9 15:46 SRX1589834_chr7_R1.fastq.gz\n",
      "-rw-rw-r--+ 1     1003 1012 92M Sep  9 15:46 SRX1589834_chr7_R2.fastq.gz\n",
      "-rw-rw-r--+ 1     1003 1012 69M Sep  9 15:47 SRX1589839_chr7_R1.fastq.gz\n",
      "-rw-rw-r--+ 1     1003 1012 70M Sep  9 15:47 SRX1589839_chr7_R2.fastq.gz\n"
     ]
    }
   ],
   "source": [
    "ls -lh /srv/data/meg-m1-a2b/blumenthal-2014/chr7/1-fastq/"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "23404525-86d6-4b34-8c90-b2acd0313100",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0021704ca4f1dd04a00d708d1311db73  /srv/data/meg-m1-a2b/blumenthal-2014/chr7/1-fastq/SRX1589831_chr7_R1.fastq.gz\n"
     ]
    }
   ],
   "source": [
    "md5sum /srv/data/meg-m1-a2b/blumenthal-2014/chr7/1-fastq/SRX1589831_chr7_R1.fastq.gz"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "fb055a50-228b-4da7-bef1-c2786338df66",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0021704ca4f1dd04a00d708d1311db73  SRX1589831_chr7_R1.fastq.gz\n",
      "558d9168b0c630a0182f60c63d931c1a  SRX1589831_chr7_R2.fastq.gz\n",
      "e8e4e8c09b0adf38c57cacb1041a9a6a  SRX1589834_chr7_R1.fastq.gz\n",
      "cf8fa4a75e6d4f2723ae8218a91e8d74  SRX1589834_chr7_R2.fastq.gz\n",
      "4765f492b28bc17892e1112576d2723c  SRX1589839_chr7_R1.fastq.gz\n",
      "1089bdb8566a7389a24f692ce95341d4  SRX1589839_chr7_R2.fastq.gz\n"
     ]
    }
   ],
   "source": [
    "head /srv/data/meg-m1-a2b/blumenthal-2014/chr7/1-fastq/md5sum.txt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "d0879f89-5a55-462a-9439-3ccd1e951982",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0021704ca4f1dd04a00d708d1311db73  /srv/data/meg-m1-a2b/blumenthal-2014/chr7/1-fastq/SRX1589831_chr7_R1.fastq.gz\n",
      "558d9168b0c630a0182f60c63d931c1a  /srv/data/meg-m1-a2b/blumenthal-2014/chr7/1-fastq/SRX1589831_chr7_R2.fastq.gz\n",
      "e8e4e8c09b0adf38c57cacb1041a9a6a  /srv/data/meg-m1-a2b/blumenthal-2014/chr7/1-fastq/SRX1589834_chr7_R1.fastq.gz\n",
      "cf8fa4a75e6d4f2723ae8218a91e8d74  /srv/data/meg-m1-a2b/blumenthal-2014/chr7/1-fastq/SRX1589834_chr7_R2.fastq.gz\n",
      "4765f492b28bc17892e1112576d2723c  /srv/data/meg-m1-a2b/blumenthal-2014/chr7/1-fastq/SRX1589839_chr7_R1.fastq.gz\n",
      "1089bdb8566a7389a24f692ce95341d4  /srv/data/meg-m1-a2b/blumenthal-2014/chr7/1-fastq/SRX1589839_chr7_R2.fastq.gz\n"
     ]
    }
   ],
   "source": [
    "md5sum /srv/data/meg-m1-a2b/blumenthal-2014/chr7/1-fastq/*.fastq.gz"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "16e58f1f",
   "metadata": {},
   "source": [
    "# 2. Contrôle qualité\n",
    "\n",
    "## FastQC\n",
    "\n",
    "La première étape d'un pipeline (RNA-seq ou autre) est de vérifier la qualité des données. L'outil `fastqc` permet de calculer un certain nombre de statistiques sur des fichiers fastq.\n",
    "\n",
    "Commencez par jeter un œil aux différentes options du programme (en général, taper le nom du programme avec l'option `-h` ou `--help` permet de voir la liste des options)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "2fbfa846",
   "metadata": {
    "tags": [],
    "vscode": {
     "languageId": "plaintext"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "            FastQC - A high throughput sequence QC analysis tool\n",
      "\n",
      "SYNOPSIS\n",
      "\n",
      "\tfastqc seqfile1 seqfile2 .. seqfileN\n",
      "\n",
      "    fastqc [-o output dir] [--(no)extract] [-f fastq|bam|sam] \n",
      "           [-c contaminant file] seqfile1 .. seqfileN\n",
      "\n",
      "DESCRIPTION\n",
      "\n",
      "    FastQC reads a set of sequence files and produces from each one a quality\n",
      "    control report consisting of a number of different modules, each one of \n",
      "    which will help to identify a different potential type of problem in your\n",
      "    data.\n",
      "    \n",
      "    If no files to process are specified on the command line then the program\n",
      "    will start as an interactive graphical application.  If files are provided\n",
      "    on the command line then the program will run with no user interaction\n",
      "    required.  In this mode it is suitable for inclusion into a standardised\n",
      "    analysis pipeline.\n",
      "    \n",
      "    The options for the program as as follows:\n",
      "    \n",
      "    -h --help       Print this help file and exit\n",
      "    \n",
      "    -v --version    Print the version of the program and exit\n",
      "    \n",
      "    -o --outdir     Create all output files in the specified output directory.\n",
      "                    Please note that this directory must exist as the program\n",
      "                    will not create it.  If this option is not set then the \n",
      "                    output file for each sequence file is created in the same\n",
      "                    directory as the sequence file which was processed.\n",
      "                    \n",
      "    --casava        Files come from raw casava output. Files in the same sample\n",
      "                    group (differing only by the group number) will be analysed\n",
      "                    as a set rather than individually. Sequences with the filter\n",
      "                    flag set in the header will be excluded from the analysis.\n",
      "                    Files must have the same names given to them by casava\n",
      "                    (including being gzipped and ending with .gz) otherwise they\n",
      "                    won't be grouped together correctly.\n",
      "                    \n",
      "    --nano          Files come from nanopore sequences and are in fast5 format. In\n",
      "                    this mode you can pass in directories to process and the program\n",
      "                    will take in all fast5 files within those directories and produce\n",
      "                    a single output file from the sequences found in all files.                    \n",
      "                    \n",
      "    --nofilter      If running with --casava then don't remove read flagged by\n",
      "                    casava as poor quality when performing the QC analysis.\n",
      "                   \n",
      "    --extract       If set then the zipped output file will be uncompressed in\n",
      "                    the same directory after it has been created. If --delete is \n",
      "                    also specified then the zip file will be removed after the \n",
      "                    contents are unzipped. \n",
      "                    \n",
      "    -j --java       Provides the full path to the java binary you want to use to\n",
      "                    launch fastqc. If not supplied then java is assumed to be in\n",
      "                    your path.\n",
      "                   \n",
      "    --noextract     Do not uncompress the output file after creating it.  You\n",
      "                    should set this option if you do not wish to uncompress\n",
      "                    the output when running in non-interactive mode.\n",
      "                    \n",
      "    --nogroup       Disable grouping of bases for reads >50bp. All reports will\n",
      "                    show data for every base in the read.  WARNING: Using this\n",
      "                    option will cause fastqc to crash and burn if you use it on\n",
      "                    really long reads, and your plots may end up a ridiculous size.\n",
      "                    You have been warned!\n",
      "                    \n",
      "    --min_length    Sets an artificial lower limit on the length of the sequence\n",
      "                    to be shown in the report.  As long as you set this to a value\n",
      "                    greater or equal to your longest read length then this will be\n",
      "                    the sequence length used to create your read groups.  This can\n",
      "                    be useful for making directly comaparable statistics from \n",
      "                    datasets with somewhat variable read lengths.\n",
      "\n",
      "    --dup_length    Sets a length to which the sequences will be truncated when \n",
      "                    defining them to be duplicates, affecting the duplication and\n",
      "                    overrepresented sequences plot.  This can be useful if you have\n",
      "                    long reads with higher levels of miscalls, or contamination with\n",
      "                    adapter dimers containing UMI sequences.\n",
      "\n",
      "                    \n",
      "    -f --format     Bypasses the normal sequence file format detection and\n",
      "                    forces the program to use the specified format.  Valid\n",
      "                    formats are bam,sam,bam_mapped,sam_mapped and fastq\n",
      "                    \n",
      "\n",
      "    --memory        Sets the base amount of memory, in Megabytes, used to process \n",
      "                    each file.  Defaults to 512MB.  You may need to increase this if\n",
      "                    you have a file with very long sequences in it.\n",
      "                \n",
      "    --svg           Save the graphs in the report in SVG format.\n",
      "\n",
      "    -t --threads    Specifies the number of files which can be processed\n",
      "                    simultaneously.  Each thread will be allocated 250MB of\n",
      "                    memory so you shouldn't run more threads than your\n",
      "                    available memory will cope with, and not more than\n",
      "                    6 threads on a 32 bit machine\n",
      "                  \n",
      "    -c              Specifies a non-default file which contains the list of\n",
      "    --contaminants  contaminants to screen overrepresented sequences against.\n",
      "                    The file must contain sets of named contaminants in the\n",
      "                    form name[tab]sequence.  Lines prefixed with a hash will\n",
      "                    be ignored.\n",
      "\n",
      "    -a              Specifies a non-default file which contains the list of\n",
      "    --adapters      adapter sequences which will be explicity searched against\n",
      "                    the library. The file must contain sets of named adapters\n",
      "                    in the form name[tab]sequence.  Lines prefixed with a hash\n",
      "                    will be ignored.\n",
      "                    \n",
      "    -l              Specifies a non-default file which contains a set of criteria\n",
      "    --limits        which will be used to determine the warn/error limits for the\n",
      "                    various modules.  This file can also be used to selectively \n",
      "                    remove some modules from the output all together.  The format\n",
      "                    needs to mirror the default limits.txt file found in the\n",
      "                    Configuration folder.\n",
      "                    \n",
      "   -k --kmers       Specifies the length of Kmer to look for in the Kmer content\n",
      "                    module. Specified Kmer length must be between 2 and 10. Default\n",
      "                    length is 7 if not specified.\n",
      "                    \n",
      "   -q --quiet       Suppress all progress messages on stdout and only report errors.\n",
      "   \n",
      "   -d --dir         Selects a directory to be used for temporary files written when\n",
      "                    generating report images. Defaults to system temp directory if\n",
      "                    not specified.\n",
      "                    \n",
      "BUGS\n",
      "\n",
      "    Any bugs in fastqc should be reported either to simon.andrews@babraham.ac.uk\n",
      "    or in www.bioinformatics.babraham.ac.uk/bugzilla/\n",
      "                   \n",
      "    \n"
     ]
    }
   ],
   "source": [
    "fastqc -h"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f01069b9",
   "metadata": {},
   "source": [
    "On décide de faire tourner l'outils sur *tous* les fichiers fastq contenus dans le dossier contenant les données."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "ab141068-647d-4caf-9189-c172faf15138",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "application/gzip\n",
      "application/gzip\n",
      "Started analysis of SRX1589831_chr7_R1.fastq.gz\n",
      "application/gzip\n",
      "application/gzip\n",
      "application/gzip\n",
      "application/gzip\n",
      "Started analysis of SRX1589831_chr7_R2.fastq.gz\n",
      "Approx 5% complete for SRX1589831_chr7_R1.fastq.gz\n",
      "Started analysis of SRX1589834_chr7_R1.fastq.gz\n",
      "Approx 5% complete for SRX1589831_chr7_R2.fastq.gz\n",
      "Started analysis of SRX1589834_chr7_R2.fastq.gz\n",
      "Approx 10% complete for SRX1589831_chr7_R1.fastq.gz\n",
      "Started analysis of SRX1589839_chr7_R1.fastq.gz\n",
      "Approx 5% complete for SRX1589834_chr7_R1.fastq.gz\n",
      "Approx 15% complete for SRX1589831_chr7_R1.fastq.gz\n",
      "Approx 10% complete for SRX1589831_chr7_R2.fastq.gz\n",
      "Approx 5% complete for SRX1589834_chr7_R2.fastq.gz\n",
      "Approx 5% complete for SRX1589839_chr7_R1.fastq.gz\n",
      "Approx 10% complete for SRX1589834_chr7_R1.fastq.gz\n",
      "Approx 20% complete for SRX1589831_chr7_R1.fastq.gz\n",
      "Approx 15% complete for SRX1589831_chr7_R2.fastq.gz\n",
      "Approx 10% complete for SRX1589839_chr7_R1.fastq.gz\n",
      "Approx 10% complete for SRX1589834_chr7_R2.fastq.gz\n",
      "Approx 25% complete for SRX1589831_chr7_R1.fastq.gz\n",
      "Approx 20% complete for SRX1589831_chr7_R2.fastq.gz\n",
      "Approx 15% complete for SRX1589834_chr7_R1.fastq.gz\n",
      "Approx 15% complete for SRX1589839_chr7_R1.fastq.gz\n",
      "Approx 15% complete for SRX1589834_chr7_R2.fastq.gz\n",
      "Approx 30% complete for SRX1589831_chr7_R1.fastq.gz\n",
      "Approx 20% complete for SRX1589839_chr7_R1.fastq.gz\n",
      "Approx 25% complete for SRX1589831_chr7_R2.fastq.gz\n",
      "Approx 20% complete for SRX1589834_chr7_R1.fastq.gz\n",
      "Approx 20% complete for SRX1589834_chr7_R2.fastq.gz\n",
      "Approx 25% complete for SRX1589839_chr7_R1.fastq.gz\n",
      "Approx 35% complete for SRX1589831_chr7_R1.fastq.gz\n",
      "Approx 30% complete for SRX1589831_chr7_R2.fastq.gz\n",
      "Approx 25% complete for SRX1589834_chr7_R1.fastq.gz\n",
      "Approx 30% complete for SRX1589839_chr7_R1.fastq.gz\n",
      "Approx 25% complete for SRX1589834_chr7_R2.fastq.gz\n",
      "Approx 40% complete for SRX1589831_chr7_R1.fastq.gz\n",
      "Approx 35% complete for SRX1589831_chr7_R2.fastq.gz\n",
      "Approx 30% complete for SRX1589834_chr7_R1.fastq.gz\n",
      "Approx 35% complete for SRX1589839_chr7_R1.fastq.gz\n",
      "Approx 45% complete for SRX1589831_chr7_R1.fastq.gz\n",
      "Approx 40% complete for SRX1589831_chr7_R2.fastq.gz\n",
      "Approx 30% complete for SRX1589834_chr7_R2.fastq.gz\n",
      "Approx 40% complete for SRX1589839_chr7_R1.fastq.gz\n",
      "Approx 35% complete for SRX1589834_chr7_R1.fastq.gz\n",
      "Approx 50% complete for SRX1589831_chr7_R1.fastq.gz\n",
      "Approx 45% complete for SRX1589831_chr7_R2.fastq.gz\n",
      "Approx 45% complete for SRX1589839_chr7_R1.fastq.gz\n",
      "Approx 35% complete for SRX1589834_chr7_R2.fastq.gz\n",
      "Approx 40% complete for SRX1589834_chr7_R1.fastq.gz\n",
      "Approx 50% complete for SRX1589839_chr7_R1.fastq.gz\n",
      "Approx 55% complete for SRX1589831_chr7_R1.fastq.gz\n",
      "Approx 50% complete for SRX1589831_chr7_R2.fastq.gz\n",
      "Approx 40% complete for SRX1589834_chr7_R2.fastq.gz\n",
      "Approx 55% complete for SRX1589839_chr7_R1.fastq.gz\n",
      "Approx 45% complete for SRX1589834_chr7_R1.fastq.gz\n",
      "Approx 55% complete for SRX1589831_chr7_R2.fastq.gz\n",
      "Approx 60% complete for SRX1589831_chr7_R1.fastq.gz\n",
      "Approx 45% complete for SRX1589834_chr7_R2.fastq.gz\n",
      "Approx 60% complete for SRX1589839_chr7_R1.fastq.gz\n",
      "Approx 60% complete for SRX1589831_chr7_R2.fastq.gz\n",
      "Approx 50% complete for SRX1589834_chr7_R1.fastq.gz\n",
      "Approx 65% complete for SRX1589831_chr7_R1.fastq.gz\n",
      "Approx 65% complete for SRX1589839_chr7_R1.fastq.gz\n",
      "Approx 50% complete for SRX1589834_chr7_R2.fastq.gz\n",
      "Approx 65% complete for SRX1589831_chr7_R2.fastq.gz\n",
      "Approx 70% complete for SRX1589831_chr7_R1.fastq.gz\n",
      "Approx 55% complete for SRX1589834_chr7_R1.fastq.gz\n",
      "Approx 70% complete for SRX1589839_chr7_R1.fastq.gz\n",
      "Approx 55% complete for SRX1589834_chr7_R2.fastq.gz\n",
      "Approx 75% complete for SRX1589831_chr7_R1.fastq.gz\n",
      "Approx 70% complete for SRX1589831_chr7_R2.fastq.gz\n",
      "Approx 60% complete for SRX1589834_chr7_R1.fastq.gz\n",
      "Approx 75% complete for SRX1589839_chr7_R1.fastq.gz\n",
      "Approx 60% complete for SRX1589834_chr7_R2.fastq.gz\n",
      "Approx 75% complete for SRX1589831_chr7_R2.fastq.gz\n",
      "Approx 80% complete for SRX1589831_chr7_R1.fastq.gz\n",
      "Approx 80% complete for SRX1589839_chr7_R1.fastq.gz\n",
      "Approx 65% complete for SRX1589834_chr7_R1.fastq.gz\n",
      "Approx 80% complete for SRX1589831_chr7_R2.fastq.gz\n",
      "Approx 85% complete for SRX1589831_chr7_R1.fastq.gz\n",
      "Approx 85% complete for SRX1589839_chr7_R1.fastq.gz\n",
      "Approx 65% complete for SRX1589834_chr7_R2.fastq.gz\n",
      "Approx 70% complete for SRX1589834_chr7_R1.fastq.gz\n",
      "Approx 85% complete for SRX1589831_chr7_R2.fastq.gz\n",
      "Approx 90% complete for SRX1589831_chr7_R1.fastq.gz\n",
      "Approx 90% complete for SRX1589839_chr7_R1.fastq.gz\n",
      "Approx 70% complete for SRX1589834_chr7_R2.fastq.gz\n",
      "Approx 90% complete for SRX1589831_chr7_R2.fastq.gz\n",
      "Approx 75% complete for SRX1589834_chr7_R1.fastq.gz\n",
      "Approx 95% complete for SRX1589831_chr7_R1.fastq.gz\n",
      "Approx 95% complete for SRX1589839_chr7_R1.fastq.gz\n",
      "Approx 75% complete for SRX1589834_chr7_R2.fastq.gz\n",
      "Approx 95% complete for SRX1589831_chr7_R2.fastq.gz\n",
      "Analysis complete for SRX1589831_chr7_R1.fastq.gz\n",
      "Analysis complete for SRX1589839_chr7_R1.fastq.gz\n",
      "Approx 80% complete for SRX1589834_chr7_R1.fastq.gz\n",
      "Approx 80% complete for SRX1589834_chr7_R2.fastq.gz\n",
      "Analysis complete for SRX1589831_chr7_R2.fastq.gz\n",
      "Approx 85% complete for SRX1589834_chr7_R1.fastq.gz\n",
      "Approx 85% complete for SRX1589834_chr7_R2.fastq.gz\n",
      "Approx 90% complete for SRX1589834_chr7_R1.fastq.gz\n",
      "Approx 90% complete for SRX1589834_chr7_R2.fastq.gz\n",
      "Approx 95% complete for SRX1589834_chr7_R1.fastq.gz\n",
      "Analysis complete for SRX1589834_chr7_R1.fastq.gz\n",
      "Approx 95% complete for SRX1589834_chr7_R2.fastq.gz\n",
      "Analysis complete for SRX1589834_chr7_R2.fastq.gz\n",
      "Started analysis of SRX1589839_chr7_R2.fastq.gz\n",
      "Approx 5% complete for SRX1589839_chr7_R2.fastq.gz\n",
      "Approx 10% complete for SRX1589839_chr7_R2.fastq.gz\n",
      "Approx 15% complete for SRX1589839_chr7_R2.fastq.gz\n",
      "Approx 20% complete for SRX1589839_chr7_R2.fastq.gz\n",
      "Approx 25% complete for SRX1589839_chr7_R2.fastq.gz\n",
      "Approx 30% complete for SRX1589839_chr7_R2.fastq.gz\n",
      "Approx 35% complete for SRX1589839_chr7_R2.fastq.gz\n",
      "Approx 40% complete for SRX1589839_chr7_R2.fastq.gz\n",
      "Approx 45% complete for SRX1589839_chr7_R2.fastq.gz\n",
      "Approx 50% complete for SRX1589839_chr7_R2.fastq.gz\n",
      "Approx 55% complete for SRX1589839_chr7_R2.fastq.gz\n",
      "Approx 60% complete for SRX1589839_chr7_R2.fastq.gz\n",
      "Approx 65% complete for SRX1589839_chr7_R2.fastq.gz\n",
      "Approx 70% complete for SRX1589839_chr7_R2.fastq.gz\n",
      "Approx 75% complete for SRX1589839_chr7_R2.fastq.gz\n",
      "Approx 80% complete for SRX1589839_chr7_R2.fastq.gz\n",
      "Approx 85% complete for SRX1589839_chr7_R2.fastq.gz\n",
      "Approx 90% complete for SRX1589839_chr7_R2.fastq.gz\n",
      "Approx 95% complete for SRX1589839_chr7_R2.fastq.gz\n",
      "Analysis complete for SRX1589839_chr7_R2.fastq.gz\n"
     ]
    }
   ],
   "source": [
    "fastqc -t 5 -o ./Analysis_chr7/1-beforeTrimming/fastqc/ \\\n",
    "    /srv/data/meg-m1-a2b/blumenthal-2014/chr7/1-fastq/*.fastq.gz"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2740a0f2",
   "metadata": {},
   "source": [
    "**Question** :\n",
    "- Combien de fichiers de sortie *par fichier fastq* obtenez-vous ?\n",
    "- Ouvrez un fichier HTML et analyser les résultats."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "174194c3",
   "metadata": {},
   "source": [
    "## MultiQC\n",
    "\n",
    "L'outil `fastqc` produit une sortie HTML par fichier fastq. Analyser *un par un* chacun des fichiers peut être trèèèèès long, surtout quand on a beaucoup d'échantillons RNA-seq.\n",
    "L'outil `multiqc` est un outil très utile qui permet d'agréger des sorties de beaucoup d'outils différents et générer un fichier HTML.\n",
    "\n",
    "Commencez par jeter un coup d'œil aux différentes options du programme."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "773b4f4b-b71b-41e3-91e2-54f87c3be920",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "                                                                                \n",
      " \u001b[34m/\u001b[0m\u001b[32m/\u001b[0m\u001b[31m/\u001b[0m \u001b]8;id=189163;https://multiqc.info\u001b\\\u001b[1mMultiQC\u001b[0m\u001b]8;;\u001b\\ 🔍 \u001b[2m| v1.13\u001b[0m                                                         \n",
      "                                                                                \n",
      " \u001b[33mUsage:\u001b[0m \u001b[1mmultiqc\u001b[0m [\u001b[1;36mOPTIONS\u001b[0m] [\u001b[1;36mANALYSIS\u001b[0m \u001b[1;36mDIRECTORY\u001b[0m]                                  \n",
      "                                                                                \n",
      " MultiQC aggregates results from bioinformatics analyses across many samples    \n",
      " into a single report.                                                          \n",
      " \u001b[2mIt searches a given directory for analysis logs and compiles a HTML report. \u001b[0m   \n",
      " \u001b[2mIt's a general use tool, perfect for summarising the output from numerous \u001b[0m     \n",
      " \u001b[2mbioinformatics tools.\u001b[0m                                                          \n",
      " \u001b[2mTo run, supply with one or more directory to scan for analysis results. For \u001b[0m   \n",
      " \u001b[2mexample, to run in the current working directory, use '\u001b[0m\u001b[1;2;34mmultiqc .\u001b[0m\u001b[2m'\u001b[0m              \n",
      "                                                                                \n",
      "\u001b[2m╭─\u001b[0m\u001b[2m Main options \u001b[0m\u001b[2m──────────────────────────────────────────────────────────────\u001b[0m\u001b[2m─╮\u001b[0m\n",
      "\u001b[2m│\u001b[0m \u001b[1;36m--force\u001b[0m            \u001b[1;32m-f\u001b[0m  Overwrite any existing reports                        \u001b[2m│\u001b[0m\n",
      "\u001b[2m│\u001b[0m \u001b[1;36m--config\u001b[0m           \u001b[1;32m-c\u001b[0m  Specific config file to load, after those in MultiQC  \u001b[2m│\u001b[0m\n",
      "\u001b[2m│\u001b[0m                        dir / home dir / working dir.                         \u001b[2m│\u001b[0m\n",
      "\u001b[2m│\u001b[0m                        \u001b[2;33m(PATH)                                               \u001b[0m \u001b[2m│\u001b[0m\n",
      "\u001b[2m│\u001b[0m \u001b[1;36m--cl-config\u001b[0m            Specify MultiQC config YAML on the command line       \u001b[2m│\u001b[0m\n",
      "\u001b[2m│\u001b[0m                        \u001b[2;33m(TEXT)                                         \u001b[0m       \u001b[2m│\u001b[0m\n",
      "\u001b[2m│\u001b[0m \u001b[1;36m--filename\u001b[0m         \u001b[1;32m-n\u001b[0m  Report filename. Use '\u001b[3;33mstdout\u001b[0m' to print to standard    \u001b[2m│\u001b[0m\n",
      "\u001b[2m│\u001b[0m                        out.                                                  \u001b[2m│\u001b[0m\n",
      "\u001b[2m│\u001b[0m                        \u001b[2;33m(TEXT)                                               \u001b[0m \u001b[2m│\u001b[0m\n",
      "\u001b[2m│\u001b[0m \u001b[1;36m--outdir\u001b[0m           \u001b[1;32m-o\u001b[0m  Create report in the specified output directory.      \u001b[2m│\u001b[0m\n",
      "\u001b[2m│\u001b[0m                        \u001b[2;33m(TEXT)                                          \u001b[0m      \u001b[2m│\u001b[0m\n",
      "\u001b[2m│\u001b[0m \u001b[1;36m--ignore\u001b[0m           \u001b[1;32m-x\u001b[0m  Ignore analysis files \u001b[2;33m(GLOB EXPRESSION)\u001b[0m               \u001b[2m│\u001b[0m\n",
      "\u001b[2m│\u001b[0m \u001b[1;36m--ignore-samples\u001b[0m       Ignore sample names \u001b[2;33m(GLOB EXPRESSION)\u001b[0m                 \u001b[2m│\u001b[0m\n",
      "\u001b[2m│\u001b[0m \u001b[1;36m--ignore-symlinks\u001b[0m      Ignore symlinked directories and files                \u001b[2m│\u001b[0m\n",
      "\u001b[2m│\u001b[0m \u001b[1;36m--file-list\u001b[0m        \u001b[1;32m-l\u001b[0m  Supply a file containing a list of file paths to be   \u001b[2m│\u001b[0m\n",
      "\u001b[2m│\u001b[0m                        searched, one per row                                 \u001b[2m│\u001b[0m\n",
      "\u001b[2m╰──────────────────────────────────────────────────────────────────────────────╯\u001b[0m\n",
      "\u001b[2m╭─\u001b[0m\u001b[2m Choosing modules to run \u001b[0m\u001b[2m───────────────────────────────────────────────────\u001b[0m\u001b[2m─╮\u001b[0m\n",
      "\u001b[2m│\u001b[0m \u001b[1;36m--module\u001b[0m     \u001b[1;32m-m\u001b[0m  Use only this module. Can specify multiple times.           \u001b[2m│\u001b[0m\n",
      "\u001b[2m│\u001b[0m                  \u001b[2;33m(MODULE NAME)                                    \u001b[0m           \u001b[2m│\u001b[0m\n",
      "\u001b[2m│\u001b[0m \u001b[1;36m--exclude\u001b[0m    \u001b[1;32m-e\u001b[0m  Do not use this module. Can specify multiple times.         \u001b[2m│\u001b[0m\n",
      "\u001b[2m│\u001b[0m                  \u001b[2;33m(MODULE NAME)                                      \u001b[0m         \u001b[2m│\u001b[0m\n",
      "\u001b[2m│\u001b[0m \u001b[1;36m--tag\u001b[0m            Use only modules which tagged with this keyword \u001b[2;33m(TEXT)\u001b[0m      \u001b[2m│\u001b[0m\n",
      "\u001b[2m│\u001b[0m \u001b[1;36m--view-tags\u001b[0m      View the available tags and which modules they load         \u001b[2m│\u001b[0m\n",
      "\u001b[2m╰──────────────────────────────────────────────────────────────────────────────╯\u001b[0m\n",
      "\u001b[2m╭─\u001b[0m\u001b[2m Sample handling \u001b[0m\u001b[2m───────────────────────────────────────────────────────────\u001b[0m\u001b[2m─╮\u001b[0m\n",
      "\u001b[2m│\u001b[0m \u001b[1;36m--dirs\u001b[0m           \u001b[1;32m-d\u001b[0m   Prepend directory to sample names                      \u001b[2m│\u001b[0m\n",
      "\u001b[2m│\u001b[0m \u001b[1;36m--dirs-depth\u001b[0m     \u001b[1;32m-dd\u001b[0m  Prepend \u001b[3;33mn\u001b[0m directories to sample names. Negative number \u001b[2m│\u001b[0m\n",
      "\u001b[2m│\u001b[0m                       to take from start of path.                            \u001b[2m│\u001b[0m\n",
      "\u001b[2m│\u001b[0m                       \u001b[2;33m(INTEGER)                                             \u001b[0m \u001b[2m│\u001b[0m\n",
      "\u001b[2m│\u001b[0m \u001b[1;36m--fullnames\u001b[0m      \u001b[1;32m-s\u001b[0m   Do not clean the sample names \u001b[3m(leave as full file \u001b[0m     \u001b[2m│\u001b[0m\n",
      "\u001b[2m│\u001b[0m                       \u001b[3mname)\u001b[0m                                                  \u001b[2m│\u001b[0m\n",
      "\u001b[2m│\u001b[0m \u001b[1;36m--fn_as_s_name\u001b[0m        Use the log filename as the sample name                \u001b[2m│\u001b[0m\n",
      "\u001b[2m│\u001b[0m \u001b[1;36m--replace-names\u001b[0m       TSV file to rename sample names during report          \u001b[2m│\u001b[0m\n",
      "\u001b[2m│\u001b[0m                       generation                                             \u001b[2m│\u001b[0m\n",
      "\u001b[2m│\u001b[0m                       \u001b[2;33m(PATH)                                                \u001b[0m \u001b[2m│\u001b[0m\n",
      "\u001b[2m╰──────────────────────────────────────────────────────────────────────────────╯\u001b[0m\n",
      "\u001b[2m╭─\u001b[0m\u001b[2m Report customisation \u001b[0m\u001b[2m──────────────────────────────────────────────────────\u001b[0m\u001b[2m─╮\u001b[0m\n",
      "\u001b[2m│\u001b[0m \u001b[1;36m--title\u001b[0m            \u001b[1;32m-i\u001b[0m  Report title. Printed as page header, used for        \u001b[2m│\u001b[0m\n",
      "\u001b[2m│\u001b[0m                        filename if not otherwise specified.                  \u001b[2m│\u001b[0m\n",
      "\u001b[2m│\u001b[0m                        \u001b[2;33m(TEXT)                                               \u001b[0m \u001b[2m│\u001b[0m\n",
      "\u001b[2m│\u001b[0m \u001b[1;36m--comment\u001b[0m          \u001b[1;32m-b\u001b[0m  Custom comment, will be printed at the top of the     \u001b[2m│\u001b[0m\n",
      "\u001b[2m│\u001b[0m                        report.                                               \u001b[2m│\u001b[0m\n",
      "\u001b[2m│\u001b[0m                        \u001b[2;33m(TEXT)                                               \u001b[0m \u001b[2m│\u001b[0m\n",
      "\u001b[2m│\u001b[0m \u001b[1;36m--template\u001b[0m         \u001b[1;32m-t\u001b[0m  Report template to use.                               \u001b[2m│\u001b[0m\n",
      "\u001b[2m│\u001b[0m                        \u001b[2;33m(default|default_dev|gathered|geo|sections|simple)\u001b[0m    \u001b[2m│\u001b[0m\n",
      "\u001b[2m│\u001b[0m \u001b[1;36m--sample-names\u001b[0m         TSV file containing alternative sample names for      \u001b[2m│\u001b[0m\n",
      "\u001b[2m│\u001b[0m                        renaming buttons in the report                        \u001b[2m│\u001b[0m\n",
      "\u001b[2m│\u001b[0m                        \u001b[2;33m(PATH)                                               \u001b[0m \u001b[2m│\u001b[0m\n",
      "\u001b[2m│\u001b[0m \u001b[1;36m--sample-filters\u001b[0m       TSV file containing show/hide patterns for the report \u001b[2m│\u001b[0m\n",
      "\u001b[2m│\u001b[0m                        \u001b[2;33m(PATH)                                               \u001b[0m \u001b[2m│\u001b[0m\n",
      "\u001b[2m│\u001b[0m \u001b[1;36m--custom-css-file\u001b[0m      Custom CSS file to add to the final report \u001b[2;33m(PATH)\u001b[0m     \u001b[2m│\u001b[0m\n",
      "\u001b[2m╰──────────────────────────────────────────────────────────────────────────────╯\u001b[0m\n",
      "\u001b[2m╭─\u001b[0m\u001b[2m Output files \u001b[0m\u001b[2m──────────────────────────────────────────────────────────────\u001b[0m\u001b[2m─╮\u001b[0m\n",
      "\u001b[2m│\u001b[0m \u001b[1;36m--flat\u001b[0m          \u001b[1;32m-fp\u001b[0m  Use only flat plots \u001b[3m(static images)\u001b[0m                     \u001b[2m│\u001b[0m\n",
      "\u001b[2m│\u001b[0m \u001b[1;36m--interactive\u001b[0m   \u001b[1;32m-ip\u001b[0m  Use only interactive plots \u001b[3m(in-browser Javascript)\u001b[0m      \u001b[2m│\u001b[0m\n",
      "\u001b[2m│\u001b[0m \u001b[1;36m--export\u001b[0m        \u001b[1;32m-p\u001b[0m   Export plots as static images in addition to the report \u001b[2m│\u001b[0m\n",
      "\u001b[2m│\u001b[0m \u001b[1;36m--data-dir\u001b[0m           Force the parsed data directory to be created.          \u001b[2m│\u001b[0m\n",
      "\u001b[2m│\u001b[0m \u001b[1;36m--no-data-dir\u001b[0m        Prevent the parsed data directory from being created.   \u001b[2m│\u001b[0m\n",
      "\u001b[2m│\u001b[0m \u001b[1;36m--data-format\u001b[0m   \u001b[1;32m-k\u001b[0m   Output parsed data in a different format.               \u001b[2m│\u001b[0m\n",
      "\u001b[2m│\u001b[0m                      \u001b[2;33m(tsv|json|yaml)                          \u001b[0m               \u001b[2m│\u001b[0m\n",
      "\u001b[2m│\u001b[0m                      \u001b[2m[default: tsv]                           \u001b[0m               \u001b[2m│\u001b[0m\n",
      "\u001b[2m│\u001b[0m \u001b[1;36m--zip-data-dir\u001b[0m  \u001b[1;32m-z\u001b[0m   Compress the data directory.                            \u001b[2m│\u001b[0m\n",
      "\u001b[2m│\u001b[0m \u001b[1;36m--no-report\u001b[0m          Do not generate a report, only export data and plots    \u001b[2m│\u001b[0m\n",
      "\u001b[2m│\u001b[0m \u001b[1;36m--pdf\u001b[0m                Creates PDF report with the \u001b[3m'simple'\u001b[0m template. Requires \u001b[2m│\u001b[0m\n",
      "\u001b[2m│\u001b[0m                      \u001b]8;id=326299;https://pandoc.org/\u001b\\Pandoc\u001b]8;;\u001b\\ to be installed.                                 \u001b[2m│\u001b[0m\n",
      "\u001b[2m╰──────────────────────────────────────────────────────────────────────────────╯\u001b[0m\n",
      "\u001b[2m╭─\u001b[0m\u001b[2m MultiQC behaviour \u001b[0m\u001b[2m─────────────────────────────────────────────────────────\u001b[0m\u001b[2m─╮\u001b[0m\n",
      "\u001b[2m│\u001b[0m \u001b[1;36m--verbose\u001b[0m           \u001b[1;32m-v\u001b[0m  Increase output verbosity. \u001b[2;33m(INTEGER RANGE)\u001b[0m           \u001b[2m│\u001b[0m\n",
      "\u001b[2m│\u001b[0m \u001b[1;36m--quiet\u001b[0m             \u001b[1;32m-q\u001b[0m  Only show log warnings                               \u001b[2m│\u001b[0m\n",
      "\u001b[2m│\u001b[0m \u001b[1;36m--lint\u001b[0m                  Use strict linting (validation) to help code         \u001b[2m│\u001b[0m\n",
      "\u001b[2m│\u001b[0m                         development                                          \u001b[2m│\u001b[0m\n",
      "\u001b[2m│\u001b[0m \u001b[1;36m--profile-runtime\u001b[0m       Add analysis of how long MultiQC takes to run to the \u001b[2m│\u001b[0m\n",
      "\u001b[2m│\u001b[0m                         report                                               \u001b[2m│\u001b[0m\n",
      "\u001b[2m│\u001b[0m \u001b[1;36m--no-megaqc-upload\u001b[0m      Don't upload generated report to MegaQC, even if     \u001b[2m│\u001b[0m\n",
      "\u001b[2m│\u001b[0m                         MegaQC options are found                             \u001b[2m│\u001b[0m\n",
      "\u001b[2m│\u001b[0m \u001b[1;36m--no-ansi\u001b[0m               Disable coloured log output                          \u001b[2m│\u001b[0m\n",
      "\u001b[2m│\u001b[0m \u001b[1;36m--version\u001b[0m               Show the version and exit.                           \u001b[2m│\u001b[0m\n",
      "\u001b[2m│\u001b[0m \u001b[1;36m--help\u001b[0m              \u001b[1;32m-h\u001b[0m  Show this message and exit.                          \u001b[2m│\u001b[0m\n",
      "\u001b[2m╰──────────────────────────────────────────────────────────────────────────────╯\u001b[0m\n",
      "                                                                                \n",
      " See \u001b]8;id=569059;http://multiqc.info\u001b\\http://multiqc.info\u001b]8;;\u001b\\ for more details.                                      \n",
      "\n"
     ]
    }
   ],
   "source": [
    "multiqc -h"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ba6bb0d5",
   "metadata": {},
   "source": [
    "Parmi les options intéressantes, l'option `-i`permet de donner un nom au fichier de sortie. C'est très pratique pour garder une tracer de ce qu'on a fait (ça évite les noms de fichiers type `toto1.html`, `toto2.html`, etc...)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "c1494045-0ae9-414f-bdf0-25bc57896440",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "  \u001b[34m/\u001b[0m\u001b[32m/\u001b[0m\u001b[31m/\u001b[0m \u001b]8;id=526178;https://multiqc.info\u001b\\\u001b[1mMultiQC\u001b[0m\u001b]8;;\u001b\\ 🔍 \u001b[2m| v1.13\u001b[0m\n",
      "\n",
      "\u001b[34m|           multiqc\u001b[0m | \u001b[33mMultiQC Version v1.31 now available!\u001b[0m\n",
      "\u001b[34m|           multiqc\u001b[0m | Report title: GSE76872 - FastQC before trimming\n",
      "\u001b[34m|           multiqc\u001b[0m | Search path : /srv/home/yclement/meg-m1-ue5-unix2/Analysis_chr7/1-beforeTrimming/fastqc\n",
      "\u001b[2K\u001b[34m|\u001b[0m         \u001b[34msearching\u001b[0m | \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[35m100%\u001b[0m \u001b[32m12/12\u001b[0m  \u001b[0m0m  \n",
      "\u001b[?25h\u001b[34m|            fastqc\u001b[0m | Found 6 reports\n",
      "\u001b[34m|           multiqc\u001b[0m | Compressing plot data\n",
      "\u001b[34m|           multiqc\u001b[0m | Report      : Analysis_chr7/1-beforeTrimming/multiqc/GSE76872-FastQC-before-trimming_multiqc_report.html\n",
      "\u001b[34m|           multiqc\u001b[0m | Data        : Analysis_chr7/1-beforeTrimming/multiqc/GSE76872-FastQC-before-trimming_multiqc_report_data\n",
      "\u001b[34m|           multiqc\u001b[0m | MultiQC complete\n"
     ]
    }
   ],
   "source": [
    "multiqc -f -o ./Analysis_chr7/1-beforeTrimming/multiqc/ \\\n",
    "    -i \"GSE76872 - FastQC before trimming\" \\\n",
    "    --interactive \\\n",
    "    ./Analysis_chr7/1-beforeTrimming/fastqc/"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d15ed003",
   "metadata": {},
   "source": [
    "# 3. Nettoyage des reads\n",
    "\n",
    "## FastP\n",
    "\n",
    "La deuxième étape d'un pipeline d'analyse RNA-seq est de nettoyer les reads de séquençage. Voici ce qu'on cherche à nettoyer : \n",
    "- Les reads / parties de reads de mauvaise qualité\n",
    "- les partie de reads contenant des adapteurs (séquences *techniques* collées aux séquences *biologiques* qui permettent le séquençage)\n",
    "- les reads trop courts\n",
    "\n",
    "Le programme `fastp` permet de faire tout cela en une fois, de manière (quasi-)automatique.\n",
    "\n",
    "Commencez par jeter un coup d'œil aux différentes options du programme."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "97a24093-3809-4a2c-816b-9da18f09a7f9",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "usage: fastp [options] ... \n",
      "options:\n",
      "  -i, --in1                            read1 input file name (string [=])\n",
      "  -o, --out1                           read1 output file name (string [=])\n",
      "  -I, --in2                            read2 input file name (string [=])\n",
      "  -O, --out2                           read2 output file name (string [=])\n",
      "      --unpaired1                      for PE input, if read1 passed QC but read2 not, it will be written to unpaired1. Default is to discard it. (string [=])\n",
      "      --unpaired2                      for PE input, if read2 passed QC but read1 not, it will be written to unpaired2. If --unpaired2 is same as --unpaired1 (default mode), both unpaired reads will be written to this same file. (string [=])\n",
      "      --overlapped_out                 for each read pair, output the overlapped region if it has no any mismatched base. (string [=])\n",
      "      --failed_out                     specify the file to store reads that cannot pass the filters. (string [=])\n",
      "  -m, --merge                          for paired-end input, merge each pair of reads into a single read if they are overlapped. The merged reads will be written to the file given by --merged_out, the unmerged reads will be written to the files specified by --out1 and --out2. The merging mode is disabled by default.\n",
      "      --merged_out                     in the merging mode, specify the file name to store merged output, or specify --stdout to stream the merged output (string [=])\n",
      "      --include_unmerged               in the merging mode, write the unmerged or unpaired reads to the file specified by --merge. Disabled by default.\n",
      "  -6, --phred64                        indicate the input is using phred64 scoring (it'll be converted to phred33, so the output will still be phred33)\n",
      "  -z, --compression                    compression level for gzip output (1 ~ 9). 1 is fastest, 9 is smallest, default is 4. (int [=4])\n",
      "      --stdin                          input from STDIN. If the STDIN is interleaved paired-end FASTQ, please also add --interleaved_in.\n",
      "      --stdout                         stream passing-filters reads to STDOUT. This option will result in interleaved FASTQ output for paired-end output. Disabled by default.\n",
      "      --interleaved_in                 indicate that <in1> is an interleaved FASTQ which contains both read1 and read2. Disabled by default.\n",
      "      --reads_to_process               specify how many reads/pairs to be processed. Default 0 means process all reads. (int [=0])\n",
      "      --dont_overwrite                 don't overwrite existing files. Overwritting is allowed by default.\n",
      "      --fix_mgi_id                     the MGI FASTQ ID format is not compatible with many BAM operation tools, enable this option to fix it.\n",
      "  -V, --verbose                        output verbose log information (i.e. when every 1M reads are processed).\n",
      "  -A, --disable_adapter_trimming       adapter trimming is enabled by default. If this option is specified, adapter trimming is disabled\n",
      "  -a, --adapter_sequence               the adapter for read1. For SE data, if not specified, the adapter will be auto-detected. For PE data, this is used if R1/R2 are found not overlapped. (string [=auto])\n",
      "      --adapter_sequence_r2            the adapter for read2 (PE data only). This is used if R1/R2 are found not overlapped. If not specified, it will be the same as <adapter_sequence> (string [=auto])\n",
      "      --adapter_fasta                  specify a FASTA file to trim both read1 and read2 (if PE) by all the sequences in this FASTA file (string [=])\n",
      "      --detect_adapter_for_pe          by default, the auto-detection for adapter is for SE data input only, turn on this option to enable it for PE data.\n",
      "  -f, --trim_front1                    trimming how many bases in front for read1, default is 0 (int [=0])\n",
      "  -t, --trim_tail1                     trimming how many bases in tail for read1, default is 0 (int [=0])\n",
      "  -b, --max_len1                       if read1 is longer than max_len1, then trim read1 at its tail to make it as long as max_len1. Default 0 means no limitation (int [=0])\n",
      "  -F, --trim_front2                    trimming how many bases in front for read2. If it's not specified, it will follow read1's settings (int [=0])\n",
      "  -T, --trim_tail2                     trimming how many bases in tail for read2. If it's not specified, it will follow read1's settings (int [=0])\n",
      "  -B, --max_len2                       if read2 is longer than max_len2, then trim read2 at its tail to make it as long as max_len2. Default 0 means no limitation. If it's not specified, it will follow read1's settings (int [=0])\n",
      "  -D, --dedup                          enable deduplication to drop the duplicated reads/pairs\n",
      "      --dup_calc_accuracy              accuracy level to calculate duplication (1~6), higher level uses more memory (1G, 2G, 4G, 8G, 16G, 24G). Default 1 for no-dedup mode, and 3 for dedup mode. (int [=0])\n",
      "      --dont_eval_duplication          don't evaluate duplication rate to save time and use less memory.\n",
      "  -g, --trim_poly_g                    force polyG tail trimming, by default trimming is automatically enabled for Illumina NextSeq/NovaSeq data\n",
      "      --poly_g_min_len                 the minimum length to detect polyG in the read tail. 10 by default. (int [=10])\n",
      "  -G, --disable_trim_poly_g            disable polyG tail trimming, by default trimming is automatically enabled for Illumina NextSeq/NovaSeq data\n",
      "  -x, --trim_poly_x                    enable polyX trimming in 3' ends.\n",
      "      --poly_x_min_len                 the minimum length to detect polyX in the read tail. 10 by default. (int [=10])\n",
      "  -5, --cut_front                      move a sliding window from front (5') to tail, drop the bases in the window if its mean quality < threshold, stop otherwise.\n",
      "  -3, --cut_tail                       move a sliding window from tail (3') to front, drop the bases in the window if its mean quality < threshold, stop otherwise.\n",
      "  -r, --cut_right                      move a sliding window from front to tail, if meet one window with mean quality < threshold, drop the bases in the window and the right part, and then stop.\n",
      "  -W, --cut_window_size                the window size option shared by cut_front, cut_tail or cut_sliding. Range: 1~1000, default: 4 (int [=4])\n",
      "  -M, --cut_mean_quality               the mean quality requirement option shared by cut_front, cut_tail or cut_sliding. Range: 1~36 default: 20 (Q20) (int [=20])\n",
      "      --cut_front_window_size          the window size option of cut_front, default to cut_window_size if not specified (int [=4])\n",
      "      --cut_front_mean_quality         the mean quality requirement option for cut_front, default to cut_mean_quality if not specified (int [=20])\n",
      "      --cut_tail_window_size           the window size option of cut_tail, default to cut_window_size if not specified (int [=4])\n",
      "      --cut_tail_mean_quality          the mean quality requirement option for cut_tail, default to cut_mean_quality if not specified (int [=20])\n",
      "      --cut_right_window_size          the window size option of cut_right, default to cut_window_size if not specified (int [=4])\n",
      "      --cut_right_mean_quality         the mean quality requirement option for cut_right, default to cut_mean_quality if not specified (int [=20])\n",
      "  -Q, --disable_quality_filtering      quality filtering is enabled by default. If this option is specified, quality filtering is disabled\n",
      "  -q, --qualified_quality_phred        the quality value that a base is qualified. Default 15 means phred quality >=Q15 is qualified. (int [=15])\n",
      "  -u, --unqualified_percent_limit      how many percents of bases are allowed to be unqualified (0~100). Default 40 means 40% (int [=40])\n",
      "  -n, --n_base_limit                   if one read's number of N base is >n_base_limit, then this read/pair is discarded. Default is 5 (int [=5])\n",
      "  -e, --average_qual                   if one read's average quality score <avg_qual, then this read/pair is discarded. Default 0 means no requirement (int [=0])\n",
      "  -L, --disable_length_filtering       length filtering is enabled by default. If this option is specified, length filtering is disabled\n",
      "  -l, --length_required                reads shorter than length_required will be discarded, default is 15. (int [=15])\n",
      "      --length_limit                   reads longer than length_limit will be discarded, default 0 means no limitation. (int [=0])\n",
      "  -y, --low_complexity_filter          enable low complexity filter. The complexity is defined as the percentage of base that is different from its next base (base[i] != base[i+1]).\n",
      "  -Y, --complexity_threshold           the threshold for low complexity filter (0~100). Default is 30, which means 30% complexity is required. (int [=30])\n",
      "      --filter_by_index1               specify a file contains a list of barcodes of index1 to be filtered out, one barcode per line (string [=])\n",
      "      --filter_by_index2               specify a file contains a list of barcodes of index2 to be filtered out, one barcode per line (string [=])\n",
      "      --filter_by_index_threshold      the allowed difference of index barcode for index filtering, default 0 means completely identical. (int [=0])\n",
      "  -c, --correction                     enable base correction in overlapped regions (only for PE data), default is disabled\n",
      "      --overlap_len_require            the minimum length to detect overlapped region of PE reads. This will affect overlap analysis based PE merge, adapter trimming and correction. 30 by default. (int [=30])\n",
      "      --overlap_diff_limit             the maximum number of mismatched bases to detect overlapped region of PE reads. This will affect overlap analysis based PE merge, adapter trimming and correction. 5 by default. (int [=5])\n",
      "      --overlap_diff_percent_limit     the maximum percentage of mismatched bases to detect overlapped region of PE reads. This will affect overlap analysis based PE merge, adapter trimming and correction. Default 20 means 20%. (int [=20])\n",
      "  -U, --umi                            enable unique molecular identifier (UMI) preprocessing\n",
      "      --umi_loc                        specify the location of UMI, can be (index1/index2/read1/read2/per_index/per_read, default is none (string [=])\n",
      "      --umi_len                        if the UMI is in read1/read2, its length should be provided (int [=0])\n",
      "      --umi_prefix                     if specified, an underline will be used to connect prefix and UMI (i.e. prefix=UMI, UMI=AATTCG, final=UMI_AATTCG). No prefix by default (string [=])\n",
      "      --umi_skip                       if the UMI is in read1/read2, fastp can skip several bases following UMI, default is 0 (int [=0])\n",
      "  -p, --overrepresentation_analysis    enable overrepresented sequence analysis.\n",
      "  -P, --overrepresentation_sampling    one in (--overrepresentation_sampling) reads will be computed for overrepresentation analysis (1~10000), smaller is slower, default is 20. (int [=20])\n",
      "  -j, --json                           the json format report file name (string [=fastp.json])\n",
      "  -h, --html                           the html format report file name (string [=fastp.html])\n",
      "  -R, --report_title                   should be quoted with ' or \", default is \"fastp report\" (string [=fastp report])\n",
      "  -w, --thread                         worker thread number, default is 3 (int [=3])\n",
      "  -s, --split                          split output by limiting total split file number with this option (2~999), a sequential number prefix will be added to output name ( 0001.out.fq, 0002.out.fq...), disabled by default (int [=0])\n",
      "  -S, --split_by_lines                 split output by limiting lines of each file with this option(>=1000), a sequential number prefix will be added to output name ( 0001.out.fq, 0002.out.fq...), disabled by default (long [=0])\n",
      "  -d, --split_prefix_digits            the digits for the sequential number padding (1~10), default is 4, so the filename will be padded as 0001.xxx, 0 to disable padding (int [=4])\n",
      "      --cut_by_quality5                DEPRECATED, use --cut_front instead.\n",
      "      --cut_by_quality3                DEPRECATED, use --cut_tail instead.\n",
      "      --cut_by_quality_aggressive      DEPRECATED, use --cut_right instead.\n",
      "      --discard_unmerged               DEPRECATED, no effect now, see the introduction for merging.\n",
      "  -?, --help                           print this message\n"
     ]
    }
   ],
   "source": [
    "fastp --help"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c05bcc66",
   "metadata": {},
   "source": [
    "Comme vous pouvez le voir, il y a un certain nombre d'options (on ne va pas toutes les explorer).\n",
    "\n",
    "Mais avant de se lancer, posons-nous quelques questions importantes :\n",
    "- Combien d'échantillons différents avons-nous ?\n",
    "- Combien de fichier(s) fastq par échantillon avons-nous ?\n",
    "- En quoi cela va-t-il affecter nos analyses ?\n",
    "- Comment faire tourner `fastp` (ou tout autre programme) en prenant cela en compte ?"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5d16415a",
   "metadata": {},
   "source": [
    "Le fichier `samples.chr7.txt` contient la liste des différents échantillons. À votre avis, que permet de faire le code contenu dans la cellule ci-dessous ?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "e512694c-2863-4099-81e2-ddc68faefd47",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "SRX1589831\n",
      "SRX1589834\n",
      "SRX1589839\n"
     ]
    }
   ],
   "source": [
    "while read line\n",
    "do\n",
    "    echo -e $line\n",
    "done < /srv/data/meg-m1-a2b/blumenthal-2014/chr7/1-fastq/samples.chr7.txt"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "18dc4cb0",
   "metadata": {},
   "source": [
    "Chaque ligne du fichier d'entrée est donc enregistrée dans une variable (nommée `$line`) que l'on va pouvoir manipuler / utiliser.\n",
    "\n",
    "Le premier réflexe serait de simplement compléter le nom de la variable pour former un nom de fichier, comme ci-dessous.\n",
    "\n",
    "Faites-tourner la cellule pour voir si vous formez un nom de fichier fastq correct."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "73896df9-0e70-4d16-b067-301fd822a0f2",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      ".fastq.gz\n",
      ".fastq.gz\n",
      ".fastq.gz\n"
     ]
    }
   ],
   "source": [
    "while read line\n",
    "do\n",
    "    echo -e $line_chr7_R1.fastq.gz\n",
    "done < /srv/data/meg-m1-a2b/blumenthal-2014/chr7/1-fastq/samples.chr7.txt"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a02824db",
   "metadata": {},
   "source": [
    "Normalement, la cellule ci-dessus ne devrait pas fonctionner (en tout cas pas donner des noms de fichiers corrects). `BASH` permet de manipuler les variables et le texte qu'elles contiennent en encadrant le nom de la variable avec des accolades (avec le `$` à l'extérieure des accolades) : \n",
    "\n",
    "`$line` --> `${line}`\n",
    "\n",
    "Le code de la cellule ci-dessous devrait afficher des noms de fichiers corrects."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "c85978e3-00fb-41d4-b0c3-70ef2b4f1d9b",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "SRX1589831_chr7_R1.fastq.gz\n",
      "SRX1589834_chr7_R1.fastq.gz\n",
      "SRX1589839_chr7_R1.fastq.gz\n"
     ]
    }
   ],
   "source": [
    "while read line\n",
    "do\n",
    "    echo -e ${line}_chr7_R1.fastq.gz\n",
    "done < /srv/data/meg-m1-a2b/blumenthal-2014/chr7/1-fastq/samples.chr7.txt"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5ea44a74",
   "metadata": {},
   "source": [
    "On peut donc afficher en mode détaillé chaque fichier de reads forward associé à chaque échantillon :"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "5d2c55f8-8a3c-4a98-92cd-c19b6a44187e",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "-rw-rw-r--+ 1 1003 1012 80M Sep  9 15:46 /srv/data/meg-m1-a2b/blumenthal-2014/chr7/1-fastq/SRX1589831_chr7_R1.fastq.gz\n",
      "-rw-rw-r--+ 1 1003 1012 90M Sep  9 15:46 /srv/data/meg-m1-a2b/blumenthal-2014/chr7/1-fastq/SRX1589834_chr7_R1.fastq.gz\n",
      "-rw-rw-r--+ 1 1003 1012 69M Sep  9 15:47 /srv/data/meg-m1-a2b/blumenthal-2014/chr7/1-fastq/SRX1589839_chr7_R1.fastq.gz\n"
     ]
    }
   ],
   "source": [
    "while read line\n",
    "do\n",
    "    ls -alh /srv/data/meg-m1-a2b/blumenthal-2014/chr7/1-fastq/${line}_chr7_R1.fastq.gz\n",
    "done < /srv/data/meg-m1-a2b/blumenthal-2014/chr7/1-fastq/samples.chr7.txt"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "40953fb6",
   "metadata": {},
   "source": [
    "Revenons à `fastp`. Comme vous l'avez vu, il y a un très grand nombre d'options disponibles pour cet outil. Dans le code ci-dessous, voici les options que l'on utilise :\n",
    "- `--in1` & `--in2` : permet de spécifier les fichiers d'entrée *forward* et *reverse*\n",
    "- `--out1` & `--out2` : permet de spécifier les fichiers de sortie *forward* et *reverse*\n",
    "- `--json` : enregistre les informations de sortie au format *json* \n",
    "- `--json` : enregistre les informations de sortie au format *html*\n",
    "- `--cut_right` : une des façons qu'à fastp de lire les reads (ici, par la partie 3')\n",
    "- `--cut_window_size` : analyse les reads par fenêtre de *x* nucléotides\n",
    "- `--cut_mean_quality` : spécifie la qualité moyenne *minimum* dans cette fenêtre\n",
    "- `--length_required` : spécifie la longueur minimum d'un read\n",
    "- `--detect_adapter_for_pe` : détection automatique d'adapteurs pour les reads *paired-end*\n",
    "- `--dedup`: déduplication des reads (duplication PCR)\n",
    "- `--dup_calc_accuracy` : précision de détection des reads dupliqués\n",
    "- `-p` : analyse de sur-représentation\n",
    "- `-P` : échantillonage pour l'analyse de sur-représentation\n",
    "- `--thread` : spécifie le nombre de CPU avec lequel le programme va tourner\n",
    "- `--verbose` : spécifie qu'on veut plus d'informations de sortie que par défaut. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "7cc53507-b840-4030-816d-861b336b751c",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Detecting adapter sequence for read1...\n",
      "No adapter detected for read1\n",
      "\n",
      "Detecting adapter sequence for read2...\n",
      "GGATTTAGCTCAGTGGTAGAGCGCTTGCCTAGCAAGCGCAAGGCCCTGGGTTCGGTCCT\n",
      "\n",
      "[20:39:28] start to load data of read2 \n",
      "[20:39:28] start to load data of read1 \n",
      "[20:39:31] Read2: loaded 1M reads \n",
      "[20:39:31] Read1: loaded 1M reads \n",
      "[20:39:34] Read1: loading completed with 1985 packs \n",
      "[20:39:34] Read2: loading completed with 1985 packs \n",
      "[20:39:36] thread 5 data processing completed \n",
      "[20:39:36] thread 5 finished \n",
      "[20:39:36] thread 4 data processing completed \n",
      "[20:39:36] thread 4 finished \n",
      "[20:39:36] thread 2 data processing completed \n",
      "[20:39:36] thread 2 finished \n",
      "[20:39:37] thread 3 data processing completed \n",
      "[20:39:37] thread 3 finished \n",
      "[20:39:38] thread 1 data processing completed \n",
      "[20:39:38] thread 1 finished \n",
      "[20:39:38] ./Analysis_chr7/2-afterTrimming/fastp/SRX1589831_chr7_R1.fastp.fastq.gz writer finished \n",
      "[20:39:38] ./Analysis_chr7/2-afterTrimming/fastp/SRX1589831_chr7_R2.fastp.fastq.gz writer finished \n",
      "[20:39:38] start to generate reports\n",
      " \n",
      "Read1 before filtering:\n",
      "total reads: 1984129\n",
      "total bases: 96301751\n",
      "Q20 bases: 95705016(99.3803%)\n",
      "Q30 bases: 92864786(96.431%)\n",
      "\n",
      "Read2 before filtering:\n",
      "total reads: 1984129\n",
      "total bases: 95521970\n",
      "Q20 bases: 94798111(99.2422%)\n",
      "Q30 bases: 91284912(95.5643%)\n",
      "\n",
      "Read1 after filtering:\n",
      "total reads: 1934254\n",
      "total bases: 93878550\n",
      "Q20 bases: 93292788(99.376%)\n",
      "Q30 bases: 90511752(96.4137%)\n",
      "\n",
      "Read2 after filtering:\n",
      "total reads: 1934254\n",
      "total bases: 93033100\n",
      "Q20 bases: 92323584(99.2374%)\n",
      "Q30 bases: 88886395(95.5428%)\n",
      "\n",
      "Filtering result:\n",
      "reads passed filter: 3954504\n",
      "reads failed due to low quality: 4\n",
      "reads failed due to too many N: 0\n",
      "reads failed due to too short: 13750\n",
      "reads with adapter trimmed: 9330\n",
      "bases trimmed due to adapters: 379947\n",
      "\n",
      "Duplication rate: 2.40105%\n",
      "\n",
      "Insert size peak (evaluated by paired-end reads): 50\n",
      "\n",
      "JSON report: ./Analysis_chr7/2-afterTrimming/fastp/SRX1589831.json\n",
      "HTML report: ./Analysis_chr7/2-afterTrimming/fastp/SRX1589831.html\n",
      "\n",
      "fastp --in1 /srv/data/meg-m1-a2b/blumenthal-2014/chr7/1-fastq/SRX1589831_chr7_R1.fastq.gz --in2 /srv/data/meg-m1-a2b/blumenthal-2014/chr7/1-fastq/SRX1589831_chr7_R2.fastq.gz --out1 ./Analysis_chr7/2-afterTrimming/fastp/SRX1589831_chr7_R1.fastp.fastq.gz --out2 ./Analysis_chr7/2-afterTrimming/fastp/SRX1589831_chr7_R2.fastp.fastq.gz --json ./Analysis_chr7/2-afterTrimming/fastp/SRX1589831.json --html ./Analysis_chr7/2-afterTrimming/fastp/SRX1589831.html --cut_right --cut_window_size 5 --cut_mean_quality 25 --average_qual 25 --length_required 20 --detect_adapter_for_pe --dedup --dup_calc_accuracy 3 -p -P 500 --thread 5 --verbose \n",
      "fastp v0.23.1, time used: 19 seconds\n",
      "Detecting adapter sequence for read1...\n",
      "No adapter detected for read1\n",
      "\n",
      "Detecting adapter sequence for read2...\n",
      "GGATTTAGCTCAGTGGTAGAGCGCTTGCCTAGCAAGCGCAAGGCCCTGGGTTCGGTCCTC\n",
      "\n",
      "[20:39:49] start to load data of read1 \n",
      "[20:39:49] start to load data of read2 \n",
      "[20:39:51] Read2: loaded 1M reads \n",
      "[20:39:51] Read1: loaded 1M reads \n",
      "[20:39:55] Read2: loaded 2M reads \n",
      "[20:39:55] Read1: loaded 2M reads \n",
      "[20:39:57] Read1: loading completed with 2310 packs \n",
      "[20:39:57] Read2: loading completed with 2310 packs \n",
      "[20:39:57] thread 5 data processing completed \n",
      "[20:39:57] thread 5 finished \n",
      "[20:39:57] thread 1 data processing completed \n",
      "[20:39:57] thread 1 finished \n",
      "[20:39:57] thread 4 data processing completed \n",
      "[20:39:57] thread 4 finished \n",
      "[20:39:57] thread 3 data processing completed \n",
      "[20:39:57] thread 3 finished \n",
      "[20:39:58] thread 2 data processing completed \n",
      "[20:39:58] thread 2 finished \n",
      "[20:39:58] ./Analysis_chr7/2-afterTrimming/fastp/SRX1589834_chr7_R1.fastp.fastq.gz writer finished \n",
      "[20:39:58] ./Analysis_chr7/2-afterTrimming/fastp/SRX1589834_chr7_R2.fastp.fastq.gz writer finished \n",
      "[20:39:58] start to generate reports\n",
      " \n",
      "Read1 before filtering:\n",
      "total reads: 2309828\n",
      "total bases: 112852689\n",
      "Q20 bases: 112388380(99.5886%)\n",
      "Q30 bases: 109817943(97.3109%)\n",
      "\n",
      "Read2 before filtering:\n",
      "total reads: 2309828\n",
      "total bases: 112388607\n",
      "Q20 bases: 111835071(99.5075%)\n",
      "Q30 bases: 108679421(96.6997%)\n",
      "\n",
      "Read1 after filtering:\n",
      "total reads: 2227186\n",
      "total bases: 108829359\n",
      "Q20 bases: 108376643(99.584%)\n",
      "Q30 bases: 105878624(97.2887%)\n",
      "\n",
      "Read2 after filtering:\n",
      "total reads: 2227186\n",
      "total bases: 108259901\n",
      "Q20 bases: 107721372(99.5026%)\n",
      "Q30 bases: 104657637(96.6726%)\n",
      "\n",
      "Filtering result:\n",
      "reads passed filter: 4598140\n",
      "reads failed due to low quality: 0\n",
      "reads failed due to too many N: 0\n",
      "reads failed due to too short: 21516\n",
      "reads with adapter trimmed: 13624\n",
      "bases trimmed due to adapters: 602410\n",
      "\n",
      "Duplication rate: 3.4768%\n",
      "\n",
      "Insert size peak (evaluated by paired-end reads): 69\n",
      "\n",
      "JSON report: ./Analysis_chr7/2-afterTrimming/fastp/SRX1589834.json\n",
      "HTML report: ./Analysis_chr7/2-afterTrimming/fastp/SRX1589834.html\n",
      "\n",
      "fastp --in1 /srv/data/meg-m1-a2b/blumenthal-2014/chr7/1-fastq/SRX1589834_chr7_R1.fastq.gz --in2 /srv/data/meg-m1-a2b/blumenthal-2014/chr7/1-fastq/SRX1589834_chr7_R2.fastq.gz --out1 ./Analysis_chr7/2-afterTrimming/fastp/SRX1589834_chr7_R1.fastp.fastq.gz --out2 ./Analysis_chr7/2-afterTrimming/fastp/SRX1589834_chr7_R2.fastp.fastq.gz --json ./Analysis_chr7/2-afterTrimming/fastp/SRX1589834.json --html ./Analysis_chr7/2-afterTrimming/fastp/SRX1589834.html --cut_right --cut_window_size 5 --cut_mean_quality 25 --average_qual 25 --length_required 20 --detect_adapter_for_pe --dedup --dup_calc_accuracy 3 -p -P 500 --thread 5 --verbose \n",
      "fastp v0.23.1, time used: 21 seconds\n",
      "Detecting adapter sequence for read1...\n",
      "No adapter detected for read1\n",
      "\n",
      "Detecting adapter sequence for read2...\n",
      "GGATTTAGCTCAGTGGTAGAGCGCTTGCCTAGCAAGCGCAAGGCCCTGGGTTCGGTCCT\n",
      "\n",
      "[20:40:09] start to load data of read1 \n",
      "[20:40:09] start to load data of read2 \n",
      "[20:40:11] Read1: loaded 1M reads \n",
      "[20:40:11] Read2: loaded 1M reads \n",
      "[20:40:14] Read1: loading completed with 1660 packs \n",
      "[20:40:14] Read2: loading completed with 1660 packs \n",
      "[20:40:15] thread 5 data processing completed \n",
      "[20:40:15] thread 5 finished \n",
      "[20:40:15] thread 1 data processing completed \n",
      "[20:40:15] thread 1 finished \n",
      "[20:40:15] thread 4 data processing completed \n",
      "[20:40:15] thread 4 finished \n",
      "[20:40:16] thread 3 data processing completed \n",
      "[20:40:16] thread 3 finished \n",
      "[20:40:16] thread 2 data processing completed \n",
      "[20:40:16] thread 2 finished \n",
      "[20:40:16] ./Analysis_chr7/2-afterTrimming/fastp/SRX1589839_chr7_R2.fastp.fastq.gz writer finished \n",
      "[20:40:16] ./Analysis_chr7/2-afterTrimming/fastp/SRX1589839_chr7_R1.fastp.fastq.gz writer finished \n",
      "[20:40:16] start to generate reports\n",
      " \n",
      "Read1 before filtering:\n",
      "total reads: 1659848\n",
      "total bases: 80595773\n",
      "Q20 bases: 80094262(99.3777%)\n",
      "Q30 bases: 77438639(96.0828%)\n",
      "\n",
      "Read2 before filtering:\n",
      "total reads: 1659848\n",
      "total bases: 79427324\n",
      "Q20 bases: 78665017(99.0402%)\n",
      "Q30 bases: 75302511(94.8068%)\n",
      "\n",
      "Read1 after filtering:\n",
      "total reads: 1628124\n",
      "total bases: 79063795\n",
      "Q20 bases: 78569706(99.3751%)\n",
      "Q30 bases: 75958043(96.0718%)\n",
      "\n",
      "Read2 after filtering:\n",
      "total reads: 1628124\n",
      "total bases: 77848939\n",
      "Q20 bases: 77099139(99.0369%)\n",
      "Q30 bases: 73796835(94.7949%)\n",
      "\n",
      "Filtering result:\n",
      "reads passed filter: 3309884\n",
      "reads failed due to low quality: 2\n",
      "reads failed due to too many N: 0\n",
      "reads failed due to too short: 9810\n",
      "reads with adapter trimmed: 6989\n",
      "bases trimmed due to adapters: 270233\n",
      "\n",
      "Duplication rate: 1.79745%\n",
      "\n",
      "Insert size peak (evaluated by paired-end reads): 66\n",
      "\n",
      "JSON report: ./Analysis_chr7/2-afterTrimming/fastp/SRX1589839.json\n",
      "HTML report: ./Analysis_chr7/2-afterTrimming/fastp/SRX1589839.html\n",
      "\n",
      "fastp --in1 /srv/data/meg-m1-a2b/blumenthal-2014/chr7/1-fastq/SRX1589839_chr7_R1.fastq.gz --in2 /srv/data/meg-m1-a2b/blumenthal-2014/chr7/1-fastq/SRX1589839_chr7_R2.fastq.gz --out1 ./Analysis_chr7/2-afterTrimming/fastp/SRX1589839_chr7_R1.fastp.fastq.gz --out2 ./Analysis_chr7/2-afterTrimming/fastp/SRX1589839_chr7_R2.fastp.fastq.gz --json ./Analysis_chr7/2-afterTrimming/fastp/SRX1589839.json --html ./Analysis_chr7/2-afterTrimming/fastp/SRX1589839.html --cut_right --cut_window_size 5 --cut_mean_quality 25 --average_qual 25 --length_required 20 --detect_adapter_for_pe --dedup --dup_calc_accuracy 3 -p -P 500 --thread 5 --verbose \n",
      "fastp v0.23.1, time used: 18 seconds\n"
     ]
    }
   ],
   "source": [
    "while read line\n",
    "do\n",
    "    fastp --in1 /srv/data/meg-m1-a2b/blumenthal-2014/chr7/1-fastq/${line}_chr7_R1.fastq.gz \\\n",
    "    --in2 /srv/data/meg-m1-a2b/blumenthal-2014/chr7/1-fastq/${line}_chr7_R2.fastq.gz \\\n",
    "    --out1 ./Analysis_chr7/2-afterTrimming/fastp/${line}_chr7_R1.fastp.fastq.gz \\\n",
    "    --out2 ./Analysis_chr7/2-afterTrimming/fastp/${line}_chr7_R2.fastp.fastq.gz \\\n",
    "    --json ./Analysis_chr7/2-afterTrimming/fastp/${line}.json \\\n",
    "    --html ./Analysis_chr7/2-afterTrimming/fastp/${line}.html \\\n",
    "    --cut_right \\\n",
    "    --cut_window_size 5 \\\n",
    "    --cut_mean_quality 25 \\\n",
    "    --average_qual 25 \\\n",
    "    --length_required 20 \\\n",
    "    --detect_adapter_for_pe \\\n",
    "    --dedup \\\n",
    "    --dup_calc_accuracy 3 \\\n",
    "    -p -P 500 \\\n",
    "    --thread 5 \\\n",
    "    --verbose\n",
    "\n",
    "done < /srv/data/meg-m1-a2b/blumenthal-2014/chr7/1-fastq/samples.chr7.txt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "1531b728-8d74-4c0b-8da7-824fb1b40560",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# multiqc -f -o ./Analysis_chr7/2-afterTrimming/multiqc/ \\\n",
    "#    -i \"GSE76872 - trimming\" \\\n",
    "#    --interactive \\\n",
    "#    ./Analysis_chr7/2-afterTrimming/fastp/*.json"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "effa9016",
   "metadata": {},
   "source": [
    "## FastQC & MultiQC\n",
    "\n",
    "Une fois que `fastp` a terminé de tourner, il faut aller voir les effets de ce programme sur nos données avec `fastqc` et `multiqc`.  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "4abd2d54-4065-4783-8489-1f893001a6ad",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "application/gzip\n",
      "application/gzip\n",
      "Started analysis of SRX1589831_chr7_R1.fastp.fastq.gz\n",
      "application/gzip\n",
      "application/gzip\n",
      "application/gzip\n",
      "application/gzip\n",
      "Started analysis of SRX1589831_chr7_R2.fastp.fastq.gz\n",
      "Approx 5% complete for SRX1589831_chr7_R1.fastp.fastq.gz\n",
      "Started analysis of SRX1589834_chr7_R1.fastp.fastq.gz\n",
      "Approx 5% complete for SRX1589831_chr7_R2.fastp.fastq.gz\n",
      "Started analysis of SRX1589834_chr7_R2.fastp.fastq.gz\n",
      "Approx 10% complete for SRX1589831_chr7_R1.fastp.fastq.gz\n",
      "Approx 5% complete for SRX1589834_chr7_R1.fastp.fastq.gz\n",
      "Approx 10% complete for SRX1589831_chr7_R2.fastp.fastq.gz\n",
      "Started analysis of SRX1589839_chr7_R1.fastp.fastq.gz\n",
      "Approx 5% complete for SRX1589834_chr7_R2.fastp.fastq.gz\n",
      "Approx 15% complete for SRX1589831_chr7_R1.fastp.fastq.gz\n",
      "Approx 15% complete for SRX1589831_chr7_R2.fastp.fastq.gz\n",
      "Approx 10% complete for SRX1589834_chr7_R1.fastp.fastq.gz\n",
      "Approx 5% complete for SRX1589839_chr7_R1.fastp.fastq.gz\n",
      "Approx 10% complete for SRX1589834_chr7_R2.fastp.fastq.gz\n",
      "Approx 20% complete for SRX1589831_chr7_R1.fastp.fastq.gz\n",
      "Approx 20% complete for SRX1589831_chr7_R2.fastp.fastq.gz\n",
      "Approx 10% complete for SRX1589839_chr7_R1.fastp.fastq.gz\n",
      "Approx 15% complete for SRX1589834_chr7_R1.fastp.fastq.gz\n",
      "Approx 15% complete for SRX1589834_chr7_R2.fastp.fastq.gz\n",
      "Approx 25% complete for SRX1589831_chr7_R2.fastp.fastq.gz\n",
      "Approx 15% complete for SRX1589839_chr7_R1.fastp.fastq.gz\n",
      "Approx 25% complete for SRX1589831_chr7_R1.fastp.fastq.gz\n",
      "Approx 20% complete for SRX1589834_chr7_R1.fastp.fastq.gz\n",
      "Approx 20% complete for SRX1589839_chr7_R1.fastp.fastq.gz\n",
      "Approx 30% complete for SRX1589831_chr7_R2.fastp.fastq.gz\n",
      "Approx 20% complete for SRX1589834_chr7_R2.fastp.fastq.gz\n",
      "Approx 30% complete for SRX1589831_chr7_R1.fastp.fastq.gz\n",
      "Approx 25% complete for SRX1589834_chr7_R1.fastp.fastq.gz\n",
      "Approx 25% complete for SRX1589839_chr7_R1.fastp.fastq.gz\n",
      "Approx 35% complete for SRX1589831_chr7_R2.fastp.fastq.gz\n",
      "Approx 35% complete for SRX1589831_chr7_R1.fastp.fastq.gz\n",
      "Approx 25% complete for SRX1589834_chr7_R2.fastp.fastq.gz\n",
      "Approx 30% complete for SRX1589839_chr7_R1.fastp.fastq.gz\n",
      "Approx 30% complete for SRX1589834_chr7_R1.fastp.fastq.gz\n",
      "Approx 40% complete for SRX1589831_chr7_R2.fastp.fastq.gz\n",
      "Approx 40% complete for SRX1589831_chr7_R1.fastp.fastq.gz\n",
      "Approx 30% complete for SRX1589834_chr7_R2.fastp.fastq.gz\n",
      "Approx 35% complete for SRX1589839_chr7_R1.fastp.fastq.gz\n",
      "Approx 35% complete for SRX1589834_chr7_R1.fastp.fastq.gz\n",
      "Approx 45% complete for SRX1589831_chr7_R2.fastp.fastq.gz\n",
      "Approx 45% complete for SRX1589831_chr7_R1.fastp.fastq.gz\n",
      "Approx 40% complete for SRX1589839_chr7_R1.fastp.fastq.gz\n",
      "Approx 35% complete for SRX1589834_chr7_R2.fastp.fastq.gz\n",
      "Approx 50% complete for SRX1589831_chr7_R2.fastp.fastq.gz\n",
      "Approx 40% complete for SRX1589834_chr7_R1.fastp.fastq.gz\n",
      "Approx 50% complete for SRX1589831_chr7_R1.fastp.fastq.gz\n",
      "Approx 45% complete for SRX1589839_chr7_R1.fastp.fastq.gz\n",
      "Approx 40% complete for SRX1589834_chr7_R2.fastp.fastq.gz\n",
      "Approx 55% complete for SRX1589831_chr7_R2.fastp.fastq.gz\n",
      "Approx 45% complete for SRX1589834_chr7_R1.fastp.fastq.gz\n",
      "Approx 55% complete for SRX1589831_chr7_R1.fastp.fastq.gz\n",
      "Approx 50% complete for SRX1589839_chr7_R1.fastp.fastq.gz\n",
      "Approx 45% complete for SRX1589834_chr7_R2.fastp.fastq.gz\n",
      "Approx 60% complete for SRX1589831_chr7_R2.fastp.fastq.gz\n",
      "Approx 60% complete for SRX1589831_chr7_R1.fastp.fastq.gz\n",
      "Approx 50% complete for SRX1589834_chr7_R1.fastp.fastq.gz\n",
      "Approx 55% complete for SRX1589839_chr7_R1.fastp.fastq.gz\n",
      "Approx 50% complete for SRX1589834_chr7_R2.fastp.fastq.gz\n",
      "Approx 65% complete for SRX1589831_chr7_R2.fastp.fastq.gz\n",
      "Approx 65% complete for SRX1589831_chr7_R1.fastp.fastq.gz\n",
      "Approx 60% complete for SRX1589839_chr7_R1.fastp.fastq.gz\n",
      "Approx 55% complete for SRX1589834_chr7_R1.fastp.fastq.gz\n",
      "Approx 55% complete for SRX1589834_chr7_R2.fastp.fastq.gz\n",
      "Approx 70% complete for SRX1589831_chr7_R2.fastp.fastq.gz\n",
      "Approx 65% complete for SRX1589839_chr7_R1.fastp.fastq.gz\n",
      "Approx 70% complete for SRX1589831_chr7_R1.fastp.fastq.gz\n",
      "Approx 60% complete for SRX1589834_chr7_R1.fastp.fastq.gz\n",
      "Approx 75% complete for SRX1589831_chr7_R2.fastp.fastq.gz\n",
      "Approx 70% complete for SRX1589839_chr7_R1.fastp.fastq.gz\n",
      "Approx 60% complete for SRX1589834_chr7_R2.fastp.fastq.gz\n",
      "Approx 75% complete for SRX1589831_chr7_R1.fastp.fastq.gz\n",
      "Approx 65% complete for SRX1589834_chr7_R1.fastp.fastq.gz\n",
      "Approx 75% complete for SRX1589839_chr7_R1.fastp.fastq.gz\n",
      "Approx 80% complete for SRX1589831_chr7_R2.fastp.fastq.gz\n",
      "Approx 80% complete for SRX1589831_chr7_R1.fastp.fastq.gz\n",
      "Approx 65% complete for SRX1589834_chr7_R2.fastp.fastq.gz\n",
      "Approx 80% complete for SRX1589839_chr7_R1.fastp.fastq.gz\n",
      "Approx 70% complete for SRX1589834_chr7_R1.fastp.fastq.gz\n",
      "Approx 85% complete for SRX1589831_chr7_R2.fastp.fastq.gz\n",
      "Approx 85% complete for SRX1589831_chr7_R1.fastp.fastq.gz\n",
      "Approx 70% complete for SRX1589834_chr7_R2.fastp.fastq.gz\n",
      "Approx 85% complete for SRX1589839_chr7_R1.fastp.fastq.gz\n",
      "Approx 90% complete for SRX1589831_chr7_R2.fastp.fastq.gz\n",
      "Approx 90% complete for SRX1589831_chr7_R1.fastp.fastq.gz\n",
      "Approx 75% complete for SRX1589834_chr7_R1.fastp.fastq.gz\n",
      "Approx 75% complete for SRX1589834_chr7_R2.fastp.fastq.gz\n",
      "Approx 90% complete for SRX1589839_chr7_R1.fastp.fastq.gz\n",
      "Approx 95% complete for SRX1589831_chr7_R2.fastp.fastq.gz\n",
      "Approx 95% complete for SRX1589831_chr7_R1.fastp.fastq.gz\n",
      "Approx 80% complete for SRX1589834_chr7_R1.fastp.fastq.gz\n",
      "Approx 95% complete for SRX1589839_chr7_R1.fastp.fastq.gz\n",
      "Approx 80% complete for SRX1589834_chr7_R2.fastp.fastq.gz\n",
      "Analysis complete for SRX1589831_chr7_R2.fastp.fastq.gz\n",
      "Analysis complete for SRX1589831_chr7_R1.fastp.fastq.gz\n",
      "Analysis complete for SRX1589839_chr7_R1.fastp.fastq.gz\n",
      "Approx 85% complete for SRX1589834_chr7_R1.fastp.fastq.gz\n",
      "Approx 85% complete for SRX1589834_chr7_R2.fastp.fastq.gz\n",
      "Approx 90% complete for SRX1589834_chr7_R1.fastp.fastq.gz\n",
      "Approx 90% complete for SRX1589834_chr7_R2.fastp.fastq.gz\n",
      "Approx 95% complete for SRX1589834_chr7_R1.fastp.fastq.gz\n",
      "Approx 95% complete for SRX1589834_chr7_R2.fastp.fastq.gz\n",
      "Analysis complete for SRX1589834_chr7_R1.fastp.fastq.gz\n",
      "Analysis complete for SRX1589834_chr7_R2.fastp.fastq.gz\n",
      "Started analysis of SRX1589839_chr7_R2.fastp.fastq.gz\n",
      "Approx 5% complete for SRX1589839_chr7_R2.fastp.fastq.gz\n",
      "Approx 10% complete for SRX1589839_chr7_R2.fastp.fastq.gz\n",
      "Approx 15% complete for SRX1589839_chr7_R2.fastp.fastq.gz\n",
      "Approx 20% complete for SRX1589839_chr7_R2.fastp.fastq.gz\n",
      "Approx 25% complete for SRX1589839_chr7_R2.fastp.fastq.gz\n",
      "Approx 30% complete for SRX1589839_chr7_R2.fastp.fastq.gz\n",
      "Approx 35% complete for SRX1589839_chr7_R2.fastp.fastq.gz\n",
      "Approx 40% complete for SRX1589839_chr7_R2.fastp.fastq.gz\n",
      "Approx 45% complete for SRX1589839_chr7_R2.fastp.fastq.gz\n",
      "Approx 50% complete for SRX1589839_chr7_R2.fastp.fastq.gz\n",
      "Approx 55% complete for SRX1589839_chr7_R2.fastp.fastq.gz\n",
      "Approx 60% complete for SRX1589839_chr7_R2.fastp.fastq.gz\n",
      "Approx 65% complete for SRX1589839_chr7_R2.fastp.fastq.gz\n",
      "Approx 70% complete for SRX1589839_chr7_R2.fastp.fastq.gz\n",
      "Approx 75% complete for SRX1589839_chr7_R2.fastp.fastq.gz\n",
      "Approx 80% complete for SRX1589839_chr7_R2.fastp.fastq.gz\n",
      "Approx 85% complete for SRX1589839_chr7_R2.fastp.fastq.gz\n",
      "Approx 90% complete for SRX1589839_chr7_R2.fastp.fastq.gz\n",
      "Approx 95% complete for SRX1589839_chr7_R2.fastp.fastq.gz\n",
      "Analysis complete for SRX1589839_chr7_R2.fastp.fastq.gz\n",
      "\n",
      "  \u001b[34m/\u001b[0m\u001b[32m/\u001b[0m\u001b[31m/\u001b[0m \u001b]8;id=66515;https://multiqc.info\u001b\\\u001b[1mMultiQC\u001b[0m\u001b]8;;\u001b\\ 🔍 \u001b[2m| v1.13\u001b[0m\n",
      "\n",
      "\u001b[34m|           multiqc\u001b[0m | \u001b[33mMultiQC Version v1.31 now available!\u001b[0m\n",
      "\u001b[34m|           multiqc\u001b[0m | Report title: GSE76872 - FastQC after trimming\n",
      "\u001b[34m|           multiqc\u001b[0m | Search path : /srv/home/yclement/meg-m1-ue5-unix2/Analysis_chr7/2-afterTrimming/fastqc\n",
      "\u001b[2K\u001b[34m|\u001b[0m         \u001b[34msearching\u001b[0m | \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[35m100%\u001b[0m \u001b[32m12/12\u001b[0m  \u001b[0m0m  \n",
      "\u001b[?25h\u001b[34m|            fastqc\u001b[0m | Found 6 reports\n",
      "\u001b[34m|           multiqc\u001b[0m | Compressing plot data\n",
      "\u001b[34m|           multiqc\u001b[0m | Report      : Analysis_chr7/2-afterTrimming/multiqc/GSE76872-FastQC-after-trimming_multiqc_report.html\n",
      "\u001b[34m|           multiqc\u001b[0m | Data        : Analysis_chr7/2-afterTrimming/multiqc/GSE76872-FastQC-after-trimming_multiqc_report_data\n",
      "\u001b[34m|           multiqc\u001b[0m | MultiQC complete\n"
     ]
    }
   ],
   "source": [
    "fastqc -t 5 -o ./Analysis_chr7/2-afterTrimming/fastqc/ \\\n",
    "    ./Analysis_chr7/2-afterTrimming/fastp/*.fastq.gz\n",
    "\n",
    "multiqc -f -o ./Analysis_chr7/2-afterTrimming/multiqc/ \\\n",
    "    -i \"GSE76872 - FastQC after trimming\" \\\n",
    "    --interactive \\\n",
    "    ./Analysis_chr7/2-afterTrimming/fastqc/"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b25538c6",
   "metadata": {},
   "source": [
    "- Que pouvez-vous dire du résultat de `fastp` ? \n",
    "- Est-ce que beaucoup de séquences ont été enlevées ?"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "10620990",
   "metadata": {},
   "source": [
    "# 4. Alignement des reads\n",
    "\n",
    "Une fois que les reads de séquençage ont été nettoyés, la troisième étape consiste à la cartographie (ou *mapping*) de ces reads *nettoyés* sur le génome d'intérêt.\n",
    "Il existe un grand nombre d'outils faisant cela (*bowtie*, *bwa*, *HISAT2*, *STAR*), ici, c'est l'outil *STAR* que l'on va utiliser. \n",
    "\n",
    "Une de ses particularités est qu'il est capable de couper les reads en plusieurs morceaux et d'aligner chacuns des morceaux.\n",
    "- En quoi cela est utile, voire nécessaire dans le cadre d'une analyse de RNA-seq ?\n",
    "- Pourquoi est-ce que d'autres outils ne faisant pas cela ne seront pas adaptés\n",
    "\n",
    "Un point important est que ces programmes ne peuvent pas travailler directement avec le génome de référence au format fasta (classique), mais on besoin d'*indéxer* le génome dans un format particulier. Cette étape a déjà été effectuée pour vous, mais gardez cette nécessité en tête.\n",
    "\n",
    "Un autre point important est que ces programmes sont très gourmands en ressources informatique (CPU, RAM) et ne sont pas fait pour tourner sur des ordinateurs personnels, mais plutôt sur des clusters de calcul.\n",
    "\n",
    "Commencez par regarder les options disponibles pour `STAR`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "5c7d371b-e3e1-4a2f-a1d4-2f8fe33283e9",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Usage: STAR  [options]... --genomeDir /path/to/genome/index/   --readFilesIn R1.fq R2.fq\n",
      "Spliced Transcripts Alignment to a Reference (c) Alexander Dobin, 2009-2022\n",
      "\n",
      "STAR version=2.7.11a\n",
      "STAR compilation time,server,dir=2023-09-15T02:58:53+0000 :/opt/conda/conda-bld/star_1694746407721/work/source\n",
      "For more details see:\n",
      "<https://github.com/alexdobin/STAR>\n",
      "<https://github.com/alexdobin/STAR/blob/master/doc/STARmanual.pdf>\n",
      "### versions\n",
      "versionGenome           2.7.4a\n",
      "    string: earliest genome index version compatible with this STAR release. Please do not change this value!\n",
      "\n",
      "### Parameter Files\n",
      "parametersFiles          -\n",
      "    string: name of a user-defined parameters file, \"-\": none. Can only be defined on the command line.\n",
      "\n",
      "### System\n",
      "sysShell            -\n",
      "    string: path to the shell binary, preferably bash, e.g. /bin/bash.\n",
      "                    - ... the default shell is executed, typically /bin/sh. This was reported to fail on some Ubuntu systems - then you need to specify path to bash.\n",
      "\n",
      "### Run Parameters\n",
      "runMode                         alignReads\n",
      "    string: type of the run.\n",
      "                                alignReads             ... map reads\n",
      "                                genomeGenerate         ... generate genome files\n",
      "                                inputAlignmentsFromBAM ... input alignments from BAM. Presently only works with --outWigType and --bamRemoveDuplicates options.\n",
      "                                liftOver               ... lift-over of GTF files (--sjdbGTFfile) between genome assemblies using chain file(s) from --genomeChainFiles.\n",
      "                                soloCellFiltering  </path/to/raw/count/dir/>   </path/to/output/prefix>    ... STARsolo cell filtering (\"calling\") without remapping, followed by the path to raw count directory and output (filtered) prefix\n",
      "\n",
      "runThreadN                      1\n",
      "    int: number of threads to run STAR\n",
      "\n",
      "runDirPerm                      User_RWX\n",
      "    string: permissions for the directories created at the run-time.\n",
      "                                User_RWX ... user-read/write/execute\n",
      "                                All_RWX  ... all-read/write/execute (same as chmod 777)\n",
      "\n",
      "runRNGseed                      777\n",
      "    int: random number generator seed.\n",
      "\n",
      "\n",
      "### Genome Parameters\n",
      "genomeDir                   ./GenomeDir/\n",
      "    string: path to the directory where genome files are stored (for --runMode alignReads) or will be generated (for --runMode generateGenome)\n",
      "\n",
      "genomeLoad                NoSharedMemory\n",
      "    string: mode of shared memory usage for the genome files. Only used with --runMode alignReads.\n",
      "                          LoadAndKeep     ... load genome into shared and keep it in memory after run\n",
      "                          LoadAndRemove   ... load genome into shared but remove it after run\n",
      "                          LoadAndExit     ... load genome into shared memory and exit, keeping the genome in memory for future runs\n",
      "                          Remove          ... do not map anything, just remove loaded genome from memory\n",
      "                          NoSharedMemory  ... do not use shared memory, each job will have its own private copy of the genome\n",
      "\n",
      "genomeFastaFiles            -\n",
      "    string(s): path(s) to the fasta files with the genome sequences, separated by spaces. These files should be plain text FASTA files, they *cannot* be zipped.\n",
      "                            Required for the genome generation (--runMode genomeGenerate). Can also be used in the mapping (--runMode alignReads) to add extra (new) sequences to the genome (e.g. spike-ins).\n",
      "\n",
      "genomeChainFiles            -\n",
      "    string: chain files for genomic liftover. Only used with --runMode liftOver .\n",
      "\n",
      "genomeFileSizes             0\n",
      "    uint(s)>0: genome files exact sizes in bytes. Typically, this should not be defined by the user.\n",
      "    \n",
      "genomeTransformOutput       None\n",
      "    string(s)               which output to transform back to original genome\n",
      "                            SAM     ... SAM/BAM alignments\n",
      "                            SJ      ... splice junctions (SJ.out.tab)\n",
      "                            Quant   ... quantifications (from --quantMode option)\n",
      "                            None    ... no transformation of the output        \n",
      "\n",
      "genomeChrSetMitochondrial   chrM M MT\n",
      "    string(s)               names of the mitochondrial chromosomes. Presently only used for STARsolo statisics output/\n",
      "\n",
      "### Genome Indexing Parameters - only used with --runMode genomeGenerate\n",
      "genomeChrBinNbits           18\n",
      "    int: =log2(chrBin), where chrBin is the size of the bins for genome storage: each chromosome will occupy an integer number of bins. For a genome with large number of contigs, it is recommended to scale this parameter as min(18, log2[max(GenomeLength/NumberOfReferences,ReadLength)]).\n",
      "\n",
      "genomeSAindexNbases         14\n",
      "    int: length (bases) of the SA pre-indexing string. Typically between 10 and 15. Longer strings will use much more memory, but allow faster searches. For small genomes, the parameter --genomeSAindexNbases must be scaled down to min(14, log2(GenomeLength)/2 - 1).\n",
      "\n",
      "genomeSAsparseD             1\n",
      "    int>0: suffux array sparsity, i.e. distance between indices: use bigger numbers to decrease needed RAM at the cost of mapping speed reduction\n",
      "\n",
      "genomeSuffixLengthMax       -1\n",
      "    int: maximum length of the suffixes, has to be longer than read length. -1 = infinite.\n",
      "    \n",
      "genomeTransformType         None\n",
      "    string: type of genome transformation\n",
      "                            None       ... no transformation\n",
      "                            Haploid    ... replace reference alleles with alternative alleles from VCF file (e.g. consensus allele)\n",
      "                            Diploid    ... create two haplotypes for each chromosome listed in VCF file, for genotypes 1|2, assumes perfect phasing (e.g. personal genome)\n",
      "\n",
      "genomeTransformVCF          -\n",
      "    string: path to VCF file for genome transformation\n",
      "\n",
      "\n",
      "    \n",
      "#####UnderDevelopment_begin : not supported - do not use\n",
      "genomeType                  Full\n",
      "    string: type of genome to generate\n",
      "                            Full                ... full (normal) genome\n",
      "                            Transcriptome       ... genome consists of transcript sequences\n",
      "                            SuperTransriptome   ... genome consists of superTranscript sequences\n",
      "#####UnderDevelopment_end\n",
      "\n",
      "# DEPRECATED: please use --genomeTransformVCF and --genomeTransformType options instead.\n",
      "#genomeConsensusFile         -\n",
      "#    string: VCF file with consensus SNPs (i.e. alternative allele is the major (AF>0.5) allele)\n",
      "# DEPRECATED \n",
      "\n",
      "\n",
      "\n",
      "### Splice Junctions Database\n",
      "sjdbFileChrStartEnd                     -\n",
      "    string(s): path to the files with genomic coordinates (chr <tab> start <tab> end <tab> strand) for the splice junction introns. Multiple files can be supplied wand will be concatenated.\n",
      "\n",
      "sjdbGTFfile                             -\n",
      "    string: path to the GTF file with annotations\n",
      "\n",
      "sjdbGTFchrPrefix                        -\n",
      "    string: prefix for chromosome names in a GTF file (e.g. 'chr' for using ENSMEBL annotations with UCSC genomes)\n",
      "\n",
      "sjdbGTFfeatureExon                      exon\n",
      "    string: feature type in GTF file to be used as exons for building transcripts\n",
      "\n",
      "sjdbGTFtagExonParentTranscript          transcript_id\n",
      "    string: GTF attribute name for parent transcript ID (default \"transcript_id\" works for GTF files)\n",
      "\n",
      "sjdbGTFtagExonParentGene                gene_id\n",
      "    string: GTF attribute name for parent gene ID (default \"gene_id\" works for GTF files)\n",
      "\n",
      "sjdbGTFtagExonParentGeneName            gene_name\n",
      "    string(s): GTF attrbute name for parent gene name\n",
      "\n",
      "sjdbGTFtagExonParentGeneType            gene_type gene_biotype\n",
      "    string(s): GTF attrbute name for parent gene type\n",
      "\n",
      "sjdbOverhang                            100\n",
      "    int>0: length of the donor/acceptor sequence on each side of the junctions, ideally = (mate_length - 1)\n",
      "\n",
      "sjdbScore                               2\n",
      "    int: extra alignment score for alignments that cross database junctions\n",
      "\n",
      "sjdbInsertSave                          Basic\n",
      "    string: which files to save when sjdb junctions are inserted on the fly at the mapping step\n",
      "                    Basic ... only small junction / transcript files\n",
      "                    All   ... all files including big Genome, SA and SAindex - this will create a complete genome directory\n",
      "\n",
      "### Variation parameters\n",
      "varVCFfile                              -\n",
      "    string: path to the VCF file that contains variation data. The 10th column should contain the genotype information, e.g. 0/1\n",
      "\n",
      "### Input Files\n",
      "inputBAMfile                -\n",
      "    string: path to BAM input file, to be used with --runMode inputAlignmentsFromBAM\n",
      "\n",
      "### Read Parameters\n",
      "readFilesType               Fastx\n",
      "    string: format of input read files\n",
      "                            Fastx       ... FASTA or FASTQ\n",
      "                            SAM SE      ... SAM or BAM single-end reads; for BAM use --readFilesCommand samtools view\n",
      "                            SAM PE      ... SAM or BAM paired-end reads; for BAM use --readFilesCommand samtools view\n",
      "                            \n",
      "readFilesSAMattrKeep        All\n",
      "    string(s): for --readFilesType SAM SE/PE, which SAM tags to keep in the output BAM, e.g.: --readFilesSAMtagsKeep RG PL\n",
      "                            All     ... keep all tags\n",
      "                            None    ... do not keep any tags\n",
      "\n",
      "readFilesIn                 Read1 Read2\n",
      "    string(s): paths to files that contain input read1 (and, if needed,  read2)\n",
      "\n",
      "readFilesManifest           -\n",
      "    string: path to the \"manifest\" file with the names of read files. The manifest file should contain 3 tab-separated columns:\n",
      "            paired-end reads: read1_file_name $tab$ read2_file_name $tab$ read_group_line.\n",
      "            single-end reads: read1_file_name $tab$ -               $tab$ read_group_line.\n",
      "            Spaces, but not tabs are allowed in file names.\n",
      "            If read_group_line does not start with ID:, it can only contain one ID field, and ID: will be added to it.\n",
      "            If read_group_line starts with ID:, it can contain several fields separated by $tab$, and all fields will be be copied verbatim into SAM @RG header line.\n",
      "\n",
      "readFilesPrefix             -\n",
      "    string: prefix for the read files names, i.e. it will be added in front of the strings in --readFilesIn\n",
      "\n",
      "readFilesCommand             -\n",
      "    string(s): command line to execute for each of the input file. This command should generate FASTA or FASTQ text and send it to stdout\n",
      "               For example: zcat - to uncompress .gz files, bzcat - to uncompress .bz2 files, etc.\n",
      "\n",
      "readMapNumber               -1\n",
      "    int: number of reads to map from the beginning of the file\n",
      "                            -1: map all reads\n",
      "\n",
      "readMatesLengthsIn          NotEqual\n",
      "    string: Equal/NotEqual - lengths of names,sequences,qualities for both mates are the same  / not the same. NotEqual is safe in all situations.\n",
      "\n",
      "readNameSeparator           /\n",
      "    string(s): character(s) separating the part of the read names that will be trimmed in output (read name after space is always trimmed)\n",
      "\n",
      "readQualityScoreBase        33\n",
      "    int>=0: number to be subtracted from the ASCII code to get Phred quality score\n",
      "\n",
      "### Read Clipping\n",
      "\n",
      "clipAdapterType             Hamming\n",
      "    string:                 adapter clipping type\n",
      "                            Hamming ... adapter clipping based on Hamming distance, with the number of mismatches controlled by --clip5pAdapterMMp\n",
      "                            CellRanger4 ... 5p and 3p adapter clipping similar to CellRanger4. Utilizes Opal package by Martin Šošić: https://github.com/Martinsos/opal\n",
      "                            None ... no adapter clipping, all other clip* parameters are disregarded\n",
      "                            \n",
      "clip3pNbases                 0\n",
      "    int(s): number(s) of bases to clip from 3p of each mate. If one value is given, it will be assumed the same for both mates.\n",
      "\n",
      "clip3pAdapterSeq            -\n",
      "    string(s): adapter sequences to clip from 3p of each mate.  If one value is given, it will be assumed the same for both mates.\n",
      "                            polyA ... polyA sequence with the length equal to read length\n",
      "\n",
      "clip3pAdapterMMp            0.1\n",
      "    double(s): max proportion of mismatches for 3p adapter clipping for each mate.  If one value is given, it will be assumed the same for both mates.\n",
      "\n",
      "clip3pAfterAdapterNbases    0\n",
      "    int(s): number of bases to clip from 3p of each mate after the adapter clipping. If one value is given, it will be assumed the same for both mates.\n",
      "\n",
      "clip5pNbases                 0\n",
      "    int(s): number(s) of bases to clip from 5p of each mate. If one value is given, it will be assumed the same for both mates.\n",
      "\n",
      "#####UnderDevelopment_begin : not supported - do not use   \n",
      "clip5pAdapterSeq            -\n",
      "    string(s): adapter sequences to clip from 5p of each mate, separated by space.\n",
      "\n",
      "clip5pAdapterMMp            0.1\n",
      "    double(s): max proportion of mismatches for 5p adapter clipping for each mate, separated by space\n",
      "\n",
      "clip5pAfterAdapterNbases    0\n",
      "    int(s): number of bases to clip from 5p of each mate after the adapter clipping, separated by space.\n",
      "#####UnderDevelopment_end\n",
      "\n",
      "### Limits\n",
      "limitGenomeGenerateRAM               31000000000\n",
      "    int>0: maximum available RAM (bytes) for genome generation\n",
      "\n",
      "limitIObufferSize                    30000000 50000000\n",
      "    int>0: max available buffers size (bytes) for input/output, per thread\n",
      "\n",
      "limitOutSAMoneReadBytes              100000\n",
      "    int>0: max size of the SAM record (bytes) for one read. Recommended value: >(2*(LengthMate1+LengthMate2+100)*outFilterMultimapNmax\n",
      "\n",
      "limitOutSJoneRead                    1000\n",
      "    int>0: max number of junctions for one read (including all multi-mappers)\n",
      "\n",
      "limitOutSJcollapsed                  1000000\n",
      "    int>0: max number of collapsed junctions\n",
      "\n",
      "limitBAMsortRAM                         0\n",
      "    int>=0: maximum available RAM (bytes) for sorting BAM. If =0, it will be set to the genome index size. 0 value can only be used with --genomeLoad NoSharedMemory option.\n",
      "\n",
      "limitSjdbInsertNsj                     1000000\n",
      "    int>=0: maximum number of junction to be inserted to the genome on the fly at the mapping stage, including those from annotations and those detected in the 1st step of the 2-pass run\n",
      "\n",
      "limitNreadsSoft                        -1\n",
      "    int: soft limit on the number of reads\n",
      "\n",
      "### Output: general\n",
      "outFileNamePrefix               ./\n",
      "    string: output files name prefix (including full or relative path). Can only be defined on the command line.\n",
      "\n",
      "outTmpDir                       -\n",
      "    string: path to a directory that will be used as temporary by STAR. All contents of this directory will be removed!\n",
      "            - the temp directory will default to outFileNamePrefix_STARtmp\n",
      "\n",
      "outTmpKeep                      None\n",
      "    string: whether to keep the tempporary files after STAR runs is finished\n",
      "                                None ... remove all temporary files\n",
      "                                All .. keep all files\n",
      "\n",
      "outStd                          Log\n",
      "    string: which output will be directed to stdout (standard out)\n",
      "                                Log                    ... log messages\n",
      "                                SAM                    ... alignments in SAM format (which normally are output to Aligned.out.sam file), normal standard output will go into Log.std.out\n",
      "                                BAM_Unsorted           ... alignments in BAM format, unsorted. Requires --outSAMtype BAM Unsorted\n",
      "                                BAM_SortedByCoordinate ... alignments in BAM format, sorted by coordinate. Requires --outSAMtype BAM SortedByCoordinate\n",
      "                                BAM_Quant              ... alignments to transcriptome in BAM format, unsorted. Requires --quantMode TranscriptomeSAM\n",
      "\n",
      "outReadsUnmapped                None\n",
      "   string: output of unmapped and partially mapped (i.e. mapped only one mate of a paired end read) reads in separate file(s).\n",
      "                                None    ... no output\n",
      "                                Fastx   ... output in separate fasta/fastq files, Unmapped.out.mate1/2\n",
      "\n",
      "outQSconversionAdd              0\n",
      "   int: add this number to the quality score (e.g. to convert from Illumina to Sanger, use -31)\n",
      "\n",
      "outMultimapperOrder             Old_2.4\n",
      "    string: order of multimapping alignments in the output files\n",
      "                                Old_2.4             ... quasi-random order used before 2.5.0\n",
      "                                Random              ... random order of alignments for each multi-mapper. Read mates (pairs) are always adjacent, all alignment for each read stay together. This option will become default in the future releases.\n",
      "\n",
      "### Output: SAM and BAM\n",
      "outSAMtype                      SAM\n",
      "    strings: type of SAM/BAM output\n",
      "                                1st word:\n",
      "                                BAM  ... output BAM without sorting\n",
      "                                SAM  ... output SAM without sorting\n",
      "                                None ... no SAM/BAM output\n",
      "                                2nd, 3rd:\n",
      "                                Unsorted           ... standard unsorted\n",
      "                                SortedByCoordinate ... sorted by coordinate. This option will allocate extra memory for sorting which can be specified by --limitBAMsortRAM.\n",
      "\n",
      "outSAMmode                      Full\n",
      "    string: mode of SAM output\n",
      "                                None ... no SAM output\n",
      "                                Full ... full SAM output\n",
      "                                NoQS ... full SAM but without quality scores\n",
      "\n",
      "outSAMstrandField               None\n",
      "    string: Cufflinks-like strand field flag\n",
      "                                None        ... not used\n",
      "                                intronMotif ... strand derived from the intron motif. This option changes the output alignments: reads with inconsistent and/or non-canonical introns are filtered out.\n",
      "\n",
      "outSAMattributes                Standard\n",
      "    string: a string of desired SAM attributes, in the order desired for the output SAM. Tags can be listed in any combination/order.\n",
      "                                ***Presets:\n",
      "                                None        ... no attributes\n",
      "                                Standard    ... NH HI AS nM\n",
      "                                All         ... NH HI AS nM NM MD jM jI MC ch                                                                    \n",
      "                                ***Alignment:\n",
      "                                NH          ... number of loci the reads maps to: =1 for unique mappers, >1 for multimappers. Standard SAM tag.\n",
      "                                HI          ... multiple alignment index, starts with --outSAMattrIHstart (=1 by default). Standard SAM tag.\n",
      "                                AS          ... local alignment score, +1/-1 for matches/mismateches, score* penalties for indels and gaps. For PE reads, total score for two mates. Stadnard SAM tag.\n",
      "                                nM          ... number of mismatches. For PE reads, sum over two mates.\n",
      "                                NM          ... edit distance to the reference (number of mismatched + inserted + deleted bases) for each mate. Standard SAM tag.\n",
      "                                MD          ... string encoding mismatched and deleted reference bases (see standard SAM specifications). Standard SAM tag.\n",
      "                                jM          ... intron motifs for all junctions (i.e. N in CIGAR): 0: non-canonical; 1: GT/AG, 2: CT/AC, 3: GC/AG, 4: CT/GC, 5: AT/AC, 6: GT/AT. If splice junctions database is used, and a junction is annotated, 20 is added to its motif value.\n",
      "                                jI          ... start and end of introns for all junctions (1-based).\n",
      "                                XS          ... alignment strand according to --outSAMstrandField.\n",
      "                                MC          ... mate's CIGAR string. Standard SAM tag.\n",
      "                                ch          ... marks all segment of all chimeric alingments for --chimOutType WithinBAM output.\n",
      "                                cN          ... number of bases clipped from the read ends: 5' and 3'\n",
      "                                ***Variation:\n",
      "                                vA          ... variant allele\n",
      "                                vG          ... genomic coordinate of the variant overlapped by the read.\n",
      "                                vW          ... 1 - alignment passes WASP filtering; 2,3,4,5,6,7 - alignment does not pass WASP filtering. Requires --waspOutputMode SAMtag.\n",
      "                                ha          ... haplotype (1/2) when mapping to the diploid genome. Requires genome generated with --genomeTransformType Diploid .                               \n",
      "                                ***STARsolo:\n",
      "                                CR CY UR UY ... sequences and quality scores of cell barcodes and UMIs for the solo* demultiplexing.\n",
      "                                GX GN       ... gene ID and gene name for unique-gene reads.\n",
      "                                gx gn       ... gene IDs and gene names for unique- and multi-gene reads.\n",
      "                                CB UB       ... error-corrected cell barcodes and UMIs for solo* demultiplexing. Requires --outSAMtype BAM SortedByCoordinate.\n",
      "                                sM          ... assessment of CB and UMI.\n",
      "                                sS          ... sequence of the entire barcode (CB,UMI,adapter).\n",
      "                                sQ          ... quality of the entire barcode.\n",
      "                                sF          ... type of feature overlap and number of features for each alignment\n",
      "                                ***Unsupported/undocumented:\n",
      "                                rB          ... alignment block read/genomic coordinates.\n",
      "                                vR          ... read coordinate of the variant.\n",
      "\n",
      "outSAMattrIHstart               1\n",
      "    int>=0:                     start value for the IH attribute. 0 may be required by some downstream software, such as Cufflinks or StringTie.\n",
      "\n",
      "outSAMunmapped                  None\n",
      "    string(s): output of unmapped reads in the SAM format\n",
      "                                1st word:\n",
      "                                None   ... no output\n",
      "                                Within ... output unmapped reads within the main SAM file (i.e. Aligned.out.sam)\n",
      "                                2nd word:\n",
      "                                KeepPairs ... record unmapped mate for each alignment, and, in case of unsorted output, keep it adjacent to its mapped mate. Only affects multi-mapping reads.\n",
      "\n",
      "outSAMorder                     Paired\n",
      "    string: type of sorting for the SAM output\n",
      "                                Paired: one mate after the other for all paired alignments\n",
      "                                PairedKeepInputOrder: one mate after the other for all paired alignments, the order is kept the same as in the input FASTQ files\n",
      "\n",
      "outSAMprimaryFlag        OneBestScore\n",
      "    string: which alignments are considered primary - all others will be marked with 0x100 bit in the FLAG\n",
      "                                OneBestScore ... only one alignment with the best score is primary\n",
      "                                AllBestScore ... all alignments with the best score are primary\n",
      "\n",
      "outSAMreadID            Standard\n",
      "    string: read ID record type\n",
      "                                Standard ... first word (until space) from the FASTx read ID line, removing /1,/2 from the end\n",
      "                                Number   ... read number (index) in the FASTx file\n",
      "\n",
      "outSAMmapqUnique        255\n",
      "    int: 0 to 255: the MAPQ value for unique mappers\n",
      "\n",
      "outSAMflagOR           0\n",
      "    int: 0 to 65535: sam FLAG will be bitwise OR'd with this value, i.e. FLAG=FLAG | outSAMflagOR. This is applied after all flags have been set by STAR, and after outSAMflagAND. Can be used to set specific bits that are not set otherwise.\n",
      "\n",
      "outSAMflagAND           65535\n",
      "    int: 0 to 65535: sam FLAG will be bitwise AND'd with this value, i.e. FLAG=FLAG & outSAMflagOR. This is applied after all flags have been set by STAR, but before outSAMflagOR. Can be used to unset specific bits that are not set otherwise.\n",
      "\n",
      "outSAMattrRGline        -\n",
      "    string(s): SAM/BAM read group line. The first word contains the read group identifier and must start with \"ID:\", e.g. --outSAMattrRGline ID:xxx CN:yy \"DS:z z z\".\n",
      "            xxx will be added as RG tag to each output alignment. Any spaces in the tag values have to be double quoted.\n",
      "            Comma separated RG lines correspons to different (comma separated) input files in --readFilesIn. Commas have to be surrounded by spaces, e.g.\n",
      "            --outSAMattrRGline ID:xxx , ID:zzz \"DS:z z\" , ID:yyy DS:yyyy\n",
      "\n",
      "outSAMheaderHD          -\n",
      "    strings: @HD (header) line of the SAM header\n",
      "\n",
      "outSAMheaderPG          -\n",
      "    strings: extra @PG (software) line of the SAM header (in addition to STAR)\n",
      "\n",
      "outSAMheaderCommentFile -\n",
      "    string: path to the file with @CO (comment) lines of the SAM header\n",
      "\n",
      "outSAMfilter            None\n",
      "    string(s): filter the output into main SAM/BAM files\n",
      "                        KeepOnlyAddedReferences ... only keep the reads for which all alignments are to the extra reference sequences added with --genomeFastaFiles at the mapping stage.\n",
      "                        KeepAllAddedReferences ...  keep all alignments to the extra reference sequences added with --genomeFastaFiles at the mapping stage.\n",
      "\n",
      "\n",
      "outSAMmultNmax          -1\n",
      "    int: max number of multiple alignments for a read that will be output to the SAM/BAM files. Note that if this value is not equal to -1, the top scoring alignment will be output first\n",
      "                        -1 ... all alignments (up to --outFilterMultimapNmax) will be output\n",
      "\n",
      "outSAMtlen              1\n",
      "    int: calculation method for the TLEN field in the SAM/BAM files\n",
      "                        1 ... leftmost base of the (+)strand mate to rightmost base of the (-)mate. (+)sign for the (+)strand mate\n",
      "                        2 ... leftmost base of any mate to rightmost base of any mate. (+)sign for the mate with the leftmost base. This is different from 1 for overlapping mates with protruding ends\n",
      "\n",
      "outBAMcompression       1\n",
      "    int: -1 to 10  BAM compression level, -1=default compression (6?), 0=no compression, 10=maximum compression\n",
      "\n",
      "outBAMsortingThreadN    0\n",
      "    int: >=0: number of threads for BAM sorting. 0 will default to min(6,--runThreadN).\n",
      "\n",
      "outBAMsortingBinsN      50\n",
      "    int: >0:  number of genome bins fo coordinate-sorting\n",
      "\n",
      "### BAM processing\n",
      "bamRemoveDuplicatesType  -\n",
      "    string: mark duplicates in the BAM file, for now only works with (i) sorted BAM fed with inputBAMfile, and (ii) for paired-end alignments only\n",
      "                        -                       ... no duplicate removal/marking\n",
      "                        UniqueIdentical         ... mark all multimappers, and duplicate unique mappers. The coordinates, FLAG, CIGAR must be identical\n",
      "                        UniqueIdenticalNotMulti  ... mark duplicate unique mappers but not multimappers.\n",
      "\n",
      "bamRemoveDuplicatesMate2basesN   0\n",
      "    int>0: number of bases from the 5' of mate 2 to use in collapsing (e.g. for RAMPAGE)\n",
      "\n",
      "### Output Wiggle\n",
      "outWigType          None\n",
      "    string(s): type of signal output, e.g. \"bedGraph\" OR \"bedGraph read1_5p\". Requires sorted BAM: --outSAMtype BAM SortedByCoordinate .\n",
      "                    1st word:\n",
      "                    None       ... no signal output\n",
      "                    bedGraph   ... bedGraph format\n",
      "                    wiggle     ... wiggle format\n",
      "                    2nd word:\n",
      "                    read1_5p   ... signal from only 5' of the 1st read, useful for CAGE/RAMPAGE etc\n",
      "                    read2      ... signal from only 2nd read\n",
      "\n",
      "outWigStrand        Stranded\n",
      "    string: strandedness of wiggle/bedGraph output\n",
      "                    Stranded   ...  separate strands, str1 and str2\n",
      "                    Unstranded ...  collapsed strands\n",
      "\n",
      "outWigReferencesPrefix    -\n",
      "    string: prefix matching reference names to include in the output wiggle file, e.g. \"chr\", default \"-\" - include all references\n",
      "\n",
      "outWigNorm              RPM\n",
      "    string: type of normalization for the signal\n",
      "                        RPM    ... reads per million of mapped reads\n",
      "                        None   ... no normalization, \"raw\" counts\n",
      "\n",
      "### Output Filtering\n",
      "outFilterType                   Normal\n",
      "    string: type of filtering\n",
      "                                Normal  ... standard filtering using only current alignment\n",
      "                                BySJout ... keep only those reads that contain junctions that passed filtering into SJ.out.tab\n",
      "\n",
      "outFilterMultimapScoreRange     1\n",
      "    int: the score range below the maximum score for multimapping alignments\n",
      "\n",
      "outFilterMultimapNmax           10\n",
      "    int: maximum number of loci the read is allowed to map to. Alignments (all of them) will be output only if the read maps to no more loci than this value.\n",
      "         Otherwise no alignments will be output, and the read will be counted as \"mapped to too many loci\" in the Log.final.out .\n",
      "\n",
      "outFilterMismatchNmax           10\n",
      "    int: alignment will be output only if it has no more mismatches than this value.\n",
      "\n",
      "outFilterMismatchNoverLmax      0.3\n",
      "    real: alignment will be output only if its ratio of mismatches to *mapped* length is less than or equal to this value.\n",
      "\n",
      "outFilterMismatchNoverReadLmax  1.0\n",
      "    real: alignment will be output only if its ratio of mismatches to *read* length is less than or equal to this value.\n",
      "\n",
      "\n",
      "outFilterScoreMin               0\n",
      "    int: alignment will be output only if its score is higher than or equal to this value.\n",
      "\n",
      "outFilterScoreMinOverLread      0.66\n",
      "    real: same as outFilterScoreMin, but  normalized to read length (sum of mates' lengths for paired-end reads)\n",
      "\n",
      "outFilterMatchNmin              0\n",
      "    int: alignment will be output only if the number of matched bases is higher than or equal to this value.\n",
      "\n",
      "outFilterMatchNminOverLread     0.66\n",
      "    real: sam as outFilterMatchNmin, but normalized to the read length (sum of mates' lengths for paired-end reads).\n",
      "\n",
      "outFilterIntronMotifs           None\n",
      "    string: filter alignment using their motifs\n",
      "                None                           ... no filtering\n",
      "                RemoveNoncanonical             ... filter out alignments that contain non-canonical junctions\n",
      "                RemoveNoncanonicalUnannotated  ... filter out alignments that contain non-canonical unannotated junctions when using annotated splice junctions database. The annotated non-canonical junctions will be kept.\n",
      "\n",
      "outFilterIntronStrands          RemoveInconsistentStrands\n",
      "    string: filter alignments\n",
      "                RemoveInconsistentStrands      ... remove alignments that have junctions with inconsistent strands\n",
      "                None                           ... no filtering\n",
      "\n",
      "### Output splice junctions (SJ.out.tab)\n",
      "outSJtype                       Standard\n",
      "    string: type of splice junction output\n",
      "                                Standard    ... standard SJ.out.tab output\n",
      "                                None        ... no splice junction output\n",
      "\n",
      "### Output Filtering: Splice Junctions\n",
      "outSJfilterReads                All\n",
      "    string: which reads to consider for collapsed splice junctions output\n",
      "                                All     ... all reads, unique- and multi-mappers\n",
      "                                Unique  ... uniquely mapping reads only\n",
      "\n",
      "outSJfilterOverhangMin          30  12  12  12\n",
      "    4 integers:    minimum overhang length for splice junctions on both sides for: (1) non-canonical motifs, (2) GT/AG and CT/AC motif, (3) GC/AG and CT/GC motif, (4) AT/AC and GT/AT motif. -1 means no output for that motif\n",
      "                                does not apply to annotated junctions\n",
      "\n",
      "outSJfilterCountUniqueMin       3   1   1   1\n",
      "    4 integers: minimum uniquely mapping read count per junction for: (1) non-canonical motifs, (2) GT/AG and CT/AC motif, (3) GC/AG and CT/GC motif, (4) AT/AC and GT/AT motif. -1 means no output for that motif\n",
      "                                Junctions are output if one of outSJfilterCountUniqueMin OR outSJfilterCountTotalMin conditions are satisfied\n",
      "                                does not apply to annotated junctions\n",
      "\n",
      "outSJfilterCountTotalMin     3   1   1   1\n",
      "    4 integers: minimum total (multi-mapping+unique) read count per junction for: (1) non-canonical motifs, (2) GT/AG and CT/AC motif, (3) GC/AG and CT/GC motif, (4) AT/AC and GT/AT motif. -1 means no output for that motif\n",
      "                                Junctions are output if one of outSJfilterCountUniqueMin OR outSJfilterCountTotalMin conditions are satisfied\n",
      "                                does not apply to annotated junctions\n",
      "\n",
      "outSJfilterDistToOtherSJmin     10  0   5   10\n",
      "    4 integers>=0: minimum allowed distance to other junctions' donor/acceptor\n",
      "                                does not apply to annotated junctions\n",
      "\n",
      "outSJfilterIntronMaxVsReadN        50000 100000 200000\n",
      "    N integers>=0: maximum gap allowed for junctions supported by 1,2,3,,,N reads\n",
      "                                i.e. by default junctions supported by 1 read can have gaps <=50000b, by 2 reads: <=100000b, by 3 reads: <=200000. by >=4 reads any gap <=alignIntronMax\n",
      "                                does not apply to annotated junctions\n",
      "\n",
      "### Scoring\n",
      "scoreGap                     0\n",
      "    int: splice junction penalty (independent on intron motif)\n",
      "\n",
      "scoreGapNoncan               -8\n",
      "    int: non-canonical junction penalty (in addition to scoreGap)\n",
      "\n",
      "scoreGapGCAG                 -4\n",
      "    GC/AG and CT/GC junction penalty (in addition to scoreGap)\n",
      "\n",
      "scoreGapATAC                 -8\n",
      "    AT/AC  and GT/AT junction penalty  (in addition to scoreGap)\n",
      "\n",
      "scoreGenomicLengthLog2scale   -0.25\n",
      "    extra score logarithmically scaled with genomic length of the alignment: scoreGenomicLengthLog2scale*log2(genomicLength)\n",
      "\n",
      "scoreDelOpen                 -2\n",
      "    deletion open penalty\n",
      "\n",
      "scoreDelBase                 -2\n",
      "    deletion extension penalty per base (in addition to scoreDelOpen)\n",
      "\n",
      "scoreInsOpen                 -2\n",
      "    insertion open penalty\n",
      "\n",
      "scoreInsBase                 -2\n",
      "    insertion extension penalty per base (in addition to scoreInsOpen)\n",
      "\n",
      "scoreStitchSJshift           1\n",
      "    maximum score reduction while searching for SJ boundaries in the stitching step\n",
      "\n",
      "\n",
      "### Alignments and Seeding\n",
      "\n",
      "seedSearchStartLmax             50\n",
      "    int>0: defines the search start point through the read - the read is split into pieces no longer than this value\n",
      "\n",
      "seedSearchStartLmaxOverLread    1.0\n",
      "    real: seedSearchStartLmax normalized to read length (sum of mates' lengths for paired-end reads)\n",
      "\n",
      "seedSearchLmax       0\n",
      "    int>=0: defines the maximum length of the seeds, if =0 seed length is not limited\n",
      "\n",
      "seedMultimapNmax      10000\n",
      "    int>0: only pieces that map fewer than this value are utilized in the stitching procedure\n",
      "\n",
      "seedPerReadNmax       1000\n",
      "    int>0: max number of seeds per read\n",
      "\n",
      "seedPerWindowNmax     50\n",
      "    int>0: max number of seeds per window\n",
      "\n",
      "seedNoneLociPerWindow    10\n",
      "    int>0: max number of one seed loci per window\n",
      "\n",
      "seedSplitMin                12\n",
      "    int>0: min length of the seed sequences split by Ns or mate gap\n",
      "\n",
      "seedMapMin              5\n",
      "    int>0: min length of seeds to be mapped\n",
      "\n",
      "alignIntronMin              21\n",
      "    minimum intron size: genomic gap is considered intron if its length>=alignIntronMin, otherwise it is considered Deletion\n",
      "\n",
      "alignIntronMax              0\n",
      "    maximum intron size, if 0, max intron size will be determined by (2^winBinNbits)*winAnchorDistNbins\n",
      "\n",
      "alignMatesGapMax            0\n",
      "    maximum gap between two mates, if 0, max intron gap will be determined by (2^winBinNbits)*winAnchorDistNbins\n",
      "\n",
      "alignSJoverhangMin          5\n",
      "    int>0: minimum overhang (i.e. block size) for spliced alignments\n",
      "\n",
      "alignSJstitchMismatchNmax   0 -1 0 0\n",
      "    4*int>=0: maximum number of mismatches for stitching of the splice junctions (-1: no limit).\n",
      "                            (1) non-canonical motifs, (2) GT/AG and CT/AC motif, (3) GC/AG and CT/GC motif, (4) AT/AC and GT/AT motif.\n",
      "\n",
      "alignSJDBoverhangMin        3\n",
      "    int>0: minimum overhang (i.e. block size) for annotated (sjdb) spliced alignments\n",
      "\n",
      "alignSplicedMateMapLmin     0\n",
      "    int>0: minimum mapped length for a read mate that is spliced\n",
      "\n",
      "alignSplicedMateMapLminOverLmate 0.66\n",
      "    real>0: alignSplicedMateMapLmin normalized to mate length\n",
      "\n",
      "alignWindowsPerReadNmax     10000\n",
      "    int>0: max number of windows per read\n",
      "\n",
      "alignTranscriptsPerWindowNmax       100\n",
      "    int>0: max number of transcripts per window\n",
      "\n",
      "alignTranscriptsPerReadNmax               10000\n",
      "    int>0: max number of different alignments per read to consider\n",
      "\n",
      "alignEndsType           Local\n",
      "    string: type of read ends alignment\n",
      "                        Local             ... standard local alignment with soft-clipping allowed\n",
      "                        EndToEnd          ... force end-to-end read alignment, do not soft-clip\n",
      "                        Extend5pOfRead1   ... fully extend only the 5p of the read1, all other ends: local alignment\n",
      "                        Extend5pOfReads12 ... fully extend only the 5p of the both read1 and read2, all other ends: local alignment\n",
      "\n",
      "alignEndsProtrude       0    ConcordantPair\n",
      "    int, string:        allow protrusion of alignment ends, i.e. start (end) of the +strand mate downstream of the start (end) of the -strand mate\n",
      "                        1st word: int: maximum number of protrusion bases allowed\n",
      "                        2nd word: string:\n",
      "                                            ConcordantPair ... report alignments with non-zero protrusion as concordant pairs\n",
      "                                            DiscordantPair ... report alignments with non-zero protrusion as discordant pairs\n",
      "\n",
      "alignSoftClipAtReferenceEnds    Yes\n",
      "    string: allow the soft-clipping of the alignments past the end of the chromosomes\n",
      "                                Yes ... allow\n",
      "                                No  ... prohibit, useful for compatibility with Cufflinks\n",
      "\n",
      "alignInsertionFlush     None\n",
      "    string: how to flush ambiguous insertion positions\n",
      "                        None    ... insertions are not flushed\n",
      "                        Right   ... insertions are flushed to the right\n",
      "\n",
      "### Paired-End reads\n",
      "peOverlapNbasesMin          0\n",
      "    int>=0:             minimum number of overlap bases to trigger mates merging and realignment. Specify >0 value to switch on the \"merginf of overlapping mates\" algorithm.\n",
      "\n",
      "peOverlapMMp                0.01\n",
      "    real, >=0 & <1:     maximum proportion of mismatched bases in the overlap area\n",
      "\n",
      "### Windows, Anchors, Binning\n",
      "\n",
      "winAnchorMultimapNmax           50\n",
      "    int>0: max number of loci anchors are allowed to map to\n",
      "\n",
      "winBinNbits                     16\n",
      "    int>0: =log2(winBin), where winBin is the size of the bin for the windows/clustering, each window will occupy an integer number of bins.\n",
      "\n",
      "winAnchorDistNbins              9\n",
      "    int>0: max number of bins between two anchors that allows aggregation of anchors into one window\n",
      "\n",
      "winFlankNbins                   4\n",
      "    int>0: log2(winFlank), where win Flank is the size of the left and right flanking regions for each window\n",
      "\n",
      "winReadCoverageRelativeMin      0.5\n",
      "    real>=0: minimum relative coverage of the read sequence by the seeds in a window, for STARlong algorithm only.\n",
      "\n",
      "winReadCoverageBasesMin      0\n",
      "    int>0: minimum number of bases covered by the seeds in a window , for STARlong algorithm only.\n",
      "\n",
      "### Chimeric Alignments\n",
      "chimOutType                 Junctions\n",
      "    string(s): type of chimeric output\n",
      "                            Junctions       ... Chimeric.out.junction\n",
      "                            SeparateSAMold  ... output old SAM into separate Chimeric.out.sam file\n",
      "                            WithinBAM       ... output into main aligned BAM files (Aligned.*.bam)\n",
      "                            WithinBAM HardClip  ... (default) hard-clipping in the CIGAR for supplemental chimeric alignments (default if no 2nd word is present)\n",
      "                            WithinBAM SoftClip  ... soft-clipping in the CIGAR for supplemental chimeric alignments\n",
      "\n",
      "chimSegmentMin              0\n",
      "    int>=0: minimum length of chimeric segment length, if ==0, no chimeric output\n",
      "\n",
      "chimScoreMin                0\n",
      "    int>=0: minimum total (summed) score of the chimeric segments\n",
      "\n",
      "chimScoreDropMax            20\n",
      "    int>=0: max drop (difference) of chimeric score (the sum of scores of all chimeric segments) from the read length\n",
      "\n",
      "chimScoreSeparation         10\n",
      "    int>=0: minimum difference (separation) between the best chimeric score and the next one\n",
      "\n",
      "chimScoreJunctionNonGTAG    -1\n",
      "    int: penalty for a non-GT/AG chimeric junction\n",
      "\n",
      "chimJunctionOverhangMin     20\n",
      "    int>=0: minimum overhang for a chimeric junction\n",
      "\n",
      "chimSegmentReadGapMax       0\n",
      "    int>=0: maximum gap in the read sequence between chimeric segments\n",
      "\n",
      "chimFilter                  banGenomicN\n",
      "    string(s): different filters for chimeric alignments\n",
      "                            None ... no filtering\n",
      "                            banGenomicN ... Ns are not allowed in the genome sequence around the chimeric junction\n",
      "\n",
      "chimMainSegmentMultNmax        10\n",
      "    int>=1: maximum number of multi-alignments for the main chimeric segment. =1 will prohibit multimapping main segments.\n",
      "\n",
      "chimMultimapNmax                    0\n",
      "    int>=0: maximum number of chimeric multi-alignments\n",
      "                                0 ... use the old scheme for chimeric detection which only considered unique alignments\n",
      "\n",
      "chimMultimapScoreRange          1\n",
      "    int>=0: the score range for multi-mapping chimeras below the best chimeric score. Only works with --chimMultimapNmax > 1\n",
      "\n",
      "chimNonchimScoreDropMin         20\n",
      "    int>=0: to trigger chimeric detection, the drop in the best non-chimeric alignment score with respect to the read length has to be greater than this value\n",
      "\n",
      "chimOutJunctionFormat           0\n",
      "    int: formatting type for the Chimeric.out.junction file\n",
      "                                0 ... no comment lines/headers\n",
      "                                1 ... comment lines at the end of the file: command line and Nreads: total, unique/multi-mapping\n",
      "\n",
      "### Quantification of Annotations\n",
      "quantMode                   -\n",
      "    string(s): types of quantification requested\n",
      "                            -                ... none\n",
      "                            TranscriptomeSAM ... output SAM/BAM alignments to transcriptome into a separate file\n",
      "                            GeneCounts       ... count reads per gene\n",
      "\n",
      "quantTranscriptomeBAMcompression    1       1\n",
      "    int: -2 to 10  transcriptome BAM compression level\n",
      "                            -2  ... no BAM output\n",
      "                            -1  ... default compression (6?)\n",
      "                             0  ... no compression\n",
      "                             10 ... maximum compression\n",
      "\n",
      "quantTranscriptomeBan       IndelSoftclipSingleend\n",
      "    string: prohibit various alignment type\n",
      "                            IndelSoftclipSingleend  ... prohibit indels, soft clipping and single-end alignments - compatible with RSEM\n",
      "                            Singleend               ... prohibit single-end alignments\n",
      "\n",
      "### 2-pass Mapping\n",
      "twopassMode                 None\n",
      "    string: 2-pass mapping mode.\n",
      "                            None        ... 1-pass mapping\n",
      "                            Basic       ... basic 2-pass mapping, with all 1st pass junctions inserted into the genome indices on the fly\n",
      "\n",
      "twopass1readsN              -1\n",
      "    int: number of reads to process for the 1st step. Use very large number (or default -1) to map all reads in the first step.\n",
      "\n",
      "\n",
      "### WASP parameters\n",
      "waspOutputMode              None\n",
      "    string: WASP allele-specific output type. This is re-implementation of the original WASP mappability filtering by Bryce van de Geijn, Graham McVicker, Yoav Gilad & Jonathan K Pritchard. Please cite the original WASP paper: Nature Methods 12, 1061–1063 (2015), https://www.nature.com/articles/nmeth.3582 .\n",
      "                            SAMtag      ... add WASP tags to the alignments that pass WASP filtering\n",
      "\n",
      "### STARsolo (single cell RNA-seq) parameters\n",
      "soloType                    None\n",
      "    string(s): type of single-cell RNA-seq\n",
      "                            CB_UMI_Simple   ... (a.k.a. Droplet) one UMI and one Cell Barcode of fixed length in read2, e.g. Drop-seq and 10X Chromium.\n",
      "                            CB_UMI_Complex  ... multiple Cell Barcodes of varying length, one UMI of fixed length and one adapter sequence of fixed length are allowed in read2 only (e.g. inDrop, ddSeq).\n",
      "                            CB_samTagOut    ... output Cell Barcode as CR and/or CB SAm tag. No UMI counting. --readFilesIn cDNA_read1 [cDNA_read2 if paired-end] CellBarcode_read . Requires --outSAMtype BAM Unsorted [and/or SortedByCoordinate]\n",
      "                            SmartSeq        ... Smart-seq: each cell in a separate FASTQ (paired- or single-end), barcodes are corresponding read-groups, no UMI sequences, alignments deduplicated according to alignment start and end (after extending soft-clipped bases)\n",
      "\n",
      "soloCBtype                  Sequence\n",
      "    string: cell barcode type\n",
      "                            Sequence: cell barcode is a sequence (standard option)\n",
      "                            String: cell barcode is an arbitrary string\n",
      "\n",
      "soloCBwhitelist             -\n",
      "    string(s): file(s) with whitelist(s) of cell barcodes. Only --soloType CB_UMI_Complex allows more than one whitelist file.\n",
      "                            None            ... no whitelist: all cell barcodes are allowed\n",
      "\n",
      "soloCBstart                 1\n",
      "    int>0: cell barcode start base\n",
      "\n",
      "soloCBlen                   16\n",
      "    int>0: cell barcode length\n",
      "\n",
      "soloUMIstart                17\n",
      "    int>0: UMI start base\n",
      "\n",
      "soloUMIlen                  10\n",
      "    int>0: UMI length\n",
      "\n",
      "soloBarcodeReadLength       1\n",
      "    int: length of the barcode read\n",
      "                            1   ... equal to sum of soloCBlen+soloUMIlen\n",
      "                            0   ... not defined, do not check\n",
      "\n",
      "soloBarcodeMate             0                            \n",
      "    int: identifies which read mate contains the barcode (CB+UMI) sequence\n",
      "                            0   ... barcode sequence is on separate read, which should always be the last file in the --readFilesIn listed\n",
      "                            1   ... barcode sequence is a part of mate 1\n",
      "                            2   ... barcode sequence is a part of mate 2\n",
      "\n",
      "soloCBposition              -\n",
      "    strings(s)              position of Cell Barcode(s) on the barcode read.\n",
      "                            Presently only works with --soloType CB_UMI_Complex, and barcodes are assumed to be on Read2.\n",
      "                            Format for each barcode: startAnchor_startPosition_endAnchor_endPosition\n",
      "                            start(end)Anchor defines the Anchor Base for the CB: 0: read start; 1: read end; 2: adapter start; 3: adapter end\n",
      "                            start(end)Position is the 0-based position with of the CB start(end) with respect to the Anchor Base\n",
      "                            String for different barcodes are separated by space.\n",
      "                            Example: inDrop (Zilionis et al, Nat. Protocols, 2017):\n",
      "                            --soloCBposition  0_0_2_-1  3_1_3_8\n",
      "\n",
      "soloUMIposition             -\n",
      "    string                  position of the UMI on the barcode read, same as soloCBposition\n",
      "                            Example: inDrop (Zilionis et al, Nat. Protocols, 2017):\n",
      "                            --soloCBposition  3_9_3_14\n",
      "\n",
      "soloAdapterSequence         -\n",
      "    string:                 adapter sequence to anchor barcodes. Only one adapter sequence is allowed.\n",
      "\n",
      "soloAdapterMismatchesNmax   1\n",
      "    int>0:                  maximum number of mismatches allowed in adapter sequence.\n",
      "\n",
      "soloCBmatchWLtype           1MM_multi\n",
      "    string:                 matching the Cell Barcodes to the WhiteList\n",
      "                            Exact                           ... only exact matches allowed\n",
      "                            1MM                             ... only one match in whitelist with 1 mismatched base allowed. Allowed CBs have to have at least one read with exact match.\n",
      "                            1MM_multi                       ... multiple matches in whitelist with 1 mismatched base allowed, posterior probability calculation is used choose one of the matches.\n",
      "                                                                Allowed CBs have to have at least one read with exact match. This option matches best with CellRanger 2.2.0\n",
      "                            1MM_multi_pseudocounts          ... same as 1MM_Multi, but pseudocounts of 1 are added to all whitelist barcodes.\n",
      "                            1MM_multi_Nbase_pseudocounts    ... same as 1MM_multi_pseudocounts, multimatching to WL is allowed for CBs with N-bases. This option matches best with CellRanger >= 3.0.0\n",
      "                            EditDist_2                    ... allow up to edit distance of 3 fpr each of the barcodes. May include one deletion + one insertion. Only works with --soloType CB_UMI_Complex. Matches to multiple passlist barcdoes are not allowed. Similar to ParseBio Split-seq pipeline. \n",
      "\n",
      "soloInputSAMattrBarcodeSeq  -\n",
      "    string(s):              when inputting reads from a SAM file (--readsFileType SAM SE/PE), these SAM attributes mark the barcode sequence (in proper order).\n",
      "                            For instance, for 10X CellRanger or STARsolo BAMs, use --soloInputSAMattrBarcodeSeq CR UR .\n",
      "                            This parameter is required when running STARsolo with input from SAM. \n",
      "    \n",
      "soloInputSAMattrBarcodeQual  -\n",
      "    string(s):              when inputting reads from a SAM file (--readsFileType SAM SE/PE), these SAM attributes mark the barcode qualities (in proper order).\n",
      "                            For instance, for 10X CellRanger or STARsolo BAMs, use --soloInputSAMattrBarcodeQual CY UY .\n",
      "                            If this parameter is '-' (default), the quality 'H' will be assigned to all bases.\n",
      "\n",
      "soloStrand                  Forward\n",
      "    string: strandedness of the solo libraries:\n",
      "                            Unstranded  ... no strand information\n",
      "                            Forward     ... read strand same as the original RNA molecule\n",
      "                            Reverse     ... read strand opposite to the original RNA molecule\n",
      "\n",
      "soloFeatures                Gene\n",
      "    string(s): genomic features for which the UMI counts per Cell Barcode are collected\n",
      "                            Gene            ... genes: reads match the gene transcript\n",
      "                            SJ              ... splice junctions: reported in SJ.out.tab\n",
      "                            GeneFull        ... full gene (pre-mRNA): count all reads overlapping genes' exons and introns\n",
      "                            GeneFull_ExonOverIntron ... full gene (pre-mRNA): count all reads overlapping genes' exons and introns: prioritize 100% overlap with exons\n",
      "                            GeneFull_Ex50pAS        ... full gene (pre-RNA): count all reads overlapping genes' exons and introns: prioritize >50% overlap with exons. Do not count reads with 100% exonic overlap in the antisense direction.\n",
      "\n",
      "#####UnderDevelopment_begin : not supported - do not use\n",
      "                            Transcript3p    ... quantification of transcript for 3' protocols\n",
      "#####UnderDevelopment_end\n",
      "\n",
      "soloMultiMappers            Unique\n",
      "    string(s): counting method for reads mapping to multiple genes           \n",
      "                            Unique     ... count only reads that map to unique genes\n",
      "                            Uniform    ... uniformly distribute multi-genic UMIs to all genes\n",
      "                            Rescue     ... distribute UMIs proportionally to unique+uniform counts (~ first iteration of EM)\n",
      "                            PropUnique ... distribute UMIs proportionally to unique mappers, if present, and uniformly if not.\n",
      "                            EM         ... multi-gene UMIs are distributed using Expectation Maximization algorithm\n",
      "\n",
      "soloUMIdedup                1MM_All\n",
      "    string(s):              type of UMI deduplication (collapsing) algorithm\n",
      "                            1MM_All                     ... all UMIs with 1 mismatch distance to each other are collapsed (i.e. counted once).\n",
      "                            1MM_Directional_UMItools    ... follows the \"directional\" method from the UMI-tools by Smith, Heger and Sudbery (Genome Research 2017).\n",
      "                            1MM_Directional             ... same as 1MM_Directional_UMItools, but with more stringent criteria for duplicate UMIs\n",
      "                            Exact                       ... only exactly matching UMIs are collapsed.\n",
      "                            NoDedup                     ... no deduplication of UMIs, count all reads.\n",
      "                            1MM_CR                      ... CellRanger2-4 algorithm for 1MM UMI collapsing.\n",
      "\n",
      "soloUMIfiltering            -\n",
      "    string(s)               type of UMI filtering (for reads uniquely mapping to genes)\n",
      "                            -                  ... basic filtering: remove UMIs with N and homopolymers (similar to CellRanger 2.2.0).\n",
      "                            MultiGeneUMI       ... basic + remove lower-count UMIs that map to more than one gene.\n",
      "                            MultiGeneUMI_All   ... basic + remove all UMIs that map to more than one gene.\n",
      "                            MultiGeneUMI_CR    ... basic + remove lower-count UMIs that map to more than one gene, matching CellRanger > 3.0.0 .\n",
      "                                                   Only works with --soloUMIdedup 1MM_CR\n",
      "                                                \n",
      "soloOutFileNames            Solo.out/          features.tsv barcodes.tsv        matrix.mtx\n",
      "    string(s)               file names for STARsolo output:\n",
      "                            file_name_prefix   gene_names   barcode_sequences   cell_feature_count_matrix\n",
      "\n",
      "soloCellFilter              CellRanger2.2 3000 0.99 10\n",
      "    string(s):              cell filtering type and parameters\n",
      "                            None            ... do not output filtered cells\n",
      "                            TopCells        ... only report top cells by UMI count, followed by the exact number of cells\n",
      "                            CellRanger2.2   ... simple filtering of CellRanger 2.2. \n",
      "                                                Can be followed by numbers: number of expected cells, robust maximum percentile for UMI count, maximum to minimum ratio for UMI count\n",
      "                                                The harcoded values are from CellRanger: nExpectedCells=3000;  maxPercentile=0.99;  maxMinRatio=10\n",
      "                            EmptyDrops_CR   ... EmptyDrops filtering in CellRanger flavor. Please cite the original EmptyDrops paper: A.T.L Lun et al, Genome Biology, 20, 63 (2019): https://genomebiology.biomedcentral.com/articles/10.1186/s13059-019-1662-y\n",
      "                                                Can be followed by 10 numeric parameters:  nExpectedCells   maxPercentile   maxMinRatio   indMin   indMax   umiMin   umiMinFracMedian   candMaxN   FDR   simN \n",
      "                                                The harcoded values are from CellRanger:             3000            0.99            10    45000    90000      500               0.01      20000  0.01  10000\n",
      "\n",
      "soloOutFormatFeaturesGeneField3    \"Gene Expression\"\n",
      "    string(s):                field 3 in the Gene features.tsv file. If \"-\", then no 3rd field is output.\n",
      "\n",
      "soloCellReadStats           None\n",
      "    string:                 Output reads statistics for each CB\n",
      "                            Standard    ... standard output\n",
      "\n",
      "#####UnderDevelopment_begin : not supported - do not use\n",
      "soloClusterCBfile           -\n",
      "    string:                 file containing the cluster information for cell barcodes, two columns: CB cluster_index. Only used with --soloFeatures Transcript3p\n",
      "#####UnderDevelopment_end\n"
     ]
    }
   ],
   "source": [
    "STAR -h"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0355efb2",
   "metadata": {},
   "source": [
    "Comme vous pouvez le voir, le nombre d'options disponibles est énorme. dans le code ci-dessous, voici les options que l'on utilise : \n",
    "- `--genomeDir` : le dossier contenant le génome indexé\n",
    "- `--runThreadN` : le nombre de CPU\n",
    "- `--readFilesIn` : les fichiers contenant les reads de séquençage\n",
    "- `--outFileNamePrefix` : le préfixe (début du nom) pour les fichiers de sortie\n",
    "- `--outSAMtype` : le format de sortie (ici au format BAM)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "319531e8-83a6-4ea1-b29e-47155b24db31",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\t/srv/conda/envs/notebook/bin/STAR-avx2 --genomeDir /srv/data/Genomes/Mmu/GRCm39/indexes_chr7_upto49bases/ --runThreadN 10 --readFilesCommand zcat --readFilesIn ./Analysis_chr7/2-afterTrimming/fastp/SRX1589831_chr7_R1.fastp.fastq.gz ./Analysis_chr7/2-afterTrimming/fastp/SRX1589831_chr7_R2.fastp.fastq.gz --outFileNamePrefix ./Analysis_chr7/3-mapping/bam/SRX1589831.chr7_ --outSAMtype BAM SortedByCoordinate\n",
      "\tSTAR version: 2.7.11a   compiled: 2023-09-15T02:58:53+0000 :/opt/conda/conda-bld/star_1694746407721/work/source\n",
      "Sep 23 20:40:41 ..... started STAR run\n",
      "Sep 23 20:40:41 ..... loading genome\n",
      "Sep 23 20:40:42 ..... started mapping\n",
      "Sep 23 20:43:38 ..... finished mapping\n",
      "Sep 23 20:43:38 ..... started sorting BAM\n",
      "Sep 23 20:43:39 ..... finished successfully\n",
      "\t/srv/conda/envs/notebook/bin/STAR-avx2 --genomeDir /srv/data/Genomes/Mmu/GRCm39/indexes_chr7_upto49bases/ --runThreadN 10 --readFilesCommand zcat --readFilesIn ./Analysis_chr7/2-afterTrimming/fastp/SRX1589834_chr7_R1.fastp.fastq.gz ./Analysis_chr7/2-afterTrimming/fastp/SRX1589834_chr7_R2.fastp.fastq.gz --outFileNamePrefix ./Analysis_chr7/3-mapping/bam/SRX1589834.chr7_ --outSAMtype BAM SortedByCoordinate\n",
      "\tSTAR version: 2.7.11a   compiled: 2023-09-15T02:58:53+0000 :/opt/conda/conda-bld/star_1694746407721/work/source\n",
      "Sep 23 20:43:40 ..... started STAR run\n",
      "Sep 23 20:43:40 ..... loading genome\n",
      "Sep 23 20:43:41 ..... started mapping\n",
      "Sep 23 20:46:37 ..... finished mapping\n",
      "Sep 23 20:46:38 ..... started sorting BAM\n",
      "Sep 23 20:46:39 ..... finished successfully\n",
      "\t/srv/conda/envs/notebook/bin/STAR-avx2 --genomeDir /srv/data/Genomes/Mmu/GRCm39/indexes_chr7_upto49bases/ --runThreadN 10 --readFilesCommand zcat --readFilesIn ./Analysis_chr7/2-afterTrimming/fastp/SRX1589839_chr7_R1.fastp.fastq.gz ./Analysis_chr7/2-afterTrimming/fastp/SRX1589839_chr7_R2.fastp.fastq.gz --outFileNamePrefix ./Analysis_chr7/3-mapping/bam/SRX1589839.chr7_ --outSAMtype BAM SortedByCoordinate\n",
      "\tSTAR version: 2.7.11a   compiled: 2023-09-15T02:58:53+0000 :/opt/conda/conda-bld/star_1694746407721/work/source\n",
      "Sep 23 20:46:39 ..... started STAR run\n",
      "Sep 23 20:46:39 ..... loading genome\n",
      "Sep 23 20:46:40 ..... started mapping\n",
      "Sep 23 20:49:33 ..... finished mapping\n",
      "Sep 23 20:49:33 ..... started sorting BAM\n",
      "Sep 23 20:49:34 ..... finished successfully\n"
     ]
    }
   ],
   "source": [
    "while read line\n",
    "do\n",
    "    STAR --genomeDir /srv/data/Genomes/Mmu/GRCm39/indexes_chr7_upto49bases/ \\\n",
    "    --runThreadN 10 \\\n",
    "    --readFilesCommand zcat \\\n",
    "    --readFilesIn ./Analysis_chr7/2-afterTrimming/fastp/${line}_chr7_R1.fastp.fastq.gz ./Analysis_chr7/2-afterTrimming/fastp/${line}_chr7_R2.fastp.fastq.gz \\\n",
    "    --outFileNamePrefix ./Analysis_chr7/3-mapping/bam/${line}.chr7_ \\\n",
    "    --outSAMtype BAM SortedByCoordinate \\\n",
    "\n",
    "done < /srv/data/meg-m1-a2b/blumenthal-2014/chr7/1-fastq/samples.chr7.txt"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ea0b1f56",
   "metadata": {},
   "source": [
    "**Questions** :\n",
    "- Combien de fichiers de sortie par échantillons produit `STAR` ?\n",
    "- Dans quel fichier peut-on trouver les informations de mapping (pourcentage de reads mappés) ?"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3811e4e3",
   "metadata": {},
   "source": [
    "## MultiQC\n",
    "\n",
    "Une fois le mapping terminé pour tous les échantillons, on peut lancer l'outil `multiqc` pour agréger les informations de mapping."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "c4297aef-4646-48ef-b0db-e38a29b63c29",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "  \u001b[34m/\u001b[0m\u001b[32m/\u001b[0m\u001b[31m/\u001b[0m \u001b]8;id=651013;https://multiqc.info\u001b\\\u001b[1mMultiQC\u001b[0m\u001b]8;;\u001b\\ 🔍 \u001b[2m| v1.13\u001b[0m\n",
      "\n",
      "\u001b[34m|           multiqc\u001b[0m | \u001b[33mMultiQC Version v1.31 now available!\u001b[0m\n",
      "\u001b[34m|           multiqc\u001b[0m | Report title: GSE76872 - Mapping\n",
      "\u001b[34m|           multiqc\u001b[0m | Search path : /srv/home/yclement/meg-m1-ue5-unix2/Analysis_chr7/3-mapping/bam\n",
      "\u001b[2K\u001b[34m|\u001b[0m         \u001b[34msearching\u001b[0m | \u001b[90m━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━\u001b[0m \u001b[35m100%\u001b[0m \u001b[32m15/15\u001b[0m  \u001b[0m0m  \n",
      "\u001b[?25h\u001b[34m|              star\u001b[0m | Found 3 reports\n",
      "\u001b[34m|           multiqc\u001b[0m | Compressing plot data\n",
      "\u001b[34m|           multiqc\u001b[0m | Report      : Analysis_chr7/3-mapping/multiqc/GSE76872-Mapping_multiqc_report.html\n",
      "\u001b[34m|           multiqc\u001b[0m | Data        : Analysis_chr7/3-mapping/multiqc/GSE76872-Mapping_multiqc_report_data\n",
      "\u001b[34m|           multiqc\u001b[0m | MultiQC complete\n"
     ]
    }
   ],
   "source": [
    "multiqc -f -o ./Analysis_chr7/3-mapping/multiqc/ \\\n",
    "    -i \"GSE76872 - Mapping\" \\\n",
    "    --interactive \\\n",
    "    ./Analysis_chr7/3-mapping/bam/"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2f69eab0",
   "metadata": {},
   "source": [
    "**Questions** : \n",
    "- Quelles informations pouvez-vous avoir sur la qualité du mapping ? Quelles informations vous semblent les plus importantes ?\n",
    "- Que pouvez-vous dire sur la qualité globale du mapping ?"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c969a2de",
   "metadata": {},
   "source": [
    "## Indexation des fichiers de mapping\n",
    "\n",
    "Une fois le mapping lancé, on lance l'outil `samtools` pour indexer les fichiers de mapping. Cette étape est nécessaire pour des outils ultérieurs (*e.x.* IGV)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "7f5c9e35-f735-4137-9177-f9e2b82ea4d5",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "samtools index -M ./Analysis_chr7/3-mapping/bam/*.bam"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Bash",
   "language": "bash",
   "name": "bash"
  },
  "language_info": {
   "codemirror_mode": "shell",
   "file_extension": ".sh",
   "mimetype": "text/x-sh",
   "name": "bash"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
